{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": "true"
   },
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Data-Wrangling-with-Pandas\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Data Wrangling with Pandas</a></div><div class=\"lev2\"><a href=\"#Date/Time-data-handling\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Date/Time data handling</a></div><div class=\"lev2\"><a href=\"#Merging-and-joining-DataFrame-objects\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Merging and joining DataFrame objects</a></div><div class=\"lev2\"><a href=\"#Concatenation\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Concatenation</a></div><div class=\"lev2\"><a href=\"#Exercise\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>Exercise</a></div><div class=\"lev2\"><a href=\"#Reshaping-DataFrame-objects\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>Reshaping DataFrame objects</a></div><div class=\"lev2\"><a href=\"#Pivoting\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>Pivoting</a></div><div class=\"lev2\"><a href=\"#Data-transformation\"><span class=\"toc-item-num\">1.7&nbsp;&nbsp;</span>Data transformation</a></div><div class=\"lev3\"><a href=\"#Dealing-with-duplicates\"><span class=\"toc-item-num\">1.7.1&nbsp;&nbsp;</span>Dealing with duplicates</a></div><div class=\"lev3\"><a href=\"#Value-replacement\"><span class=\"toc-item-num\">1.7.2&nbsp;&nbsp;</span>Value replacement</a></div><div class=\"lev3\"><a href=\"#Inidcator-variables\"><span class=\"toc-item-num\">1.7.3&nbsp;&nbsp;</span>Inidcator variables</a></div><div class=\"lev2\"><a href=\"#Categorical-Data\"><span class=\"toc-item-num\">1.8&nbsp;&nbsp;</span>Categorical Data</a></div><div class=\"lev3\"><a href=\"#Discretization\"><span class=\"toc-item-num\">1.8.1&nbsp;&nbsp;</span>Discretization</a></div><div class=\"lev3\"><a href=\"#Permutation-and-sampling\"><span class=\"toc-item-num\">1.8.2&nbsp;&nbsp;</span>Permutation and sampling</a></div><div class=\"lev2\"><a href=\"#Data-aggregation-and-GroupBy-operations\"><span class=\"toc-item-num\">1.9&nbsp;&nbsp;</span>Data aggregation and GroupBy operations</a></div><div class=\"lev3\"><a href=\"#Apply\"><span class=\"toc-item-num\">1.9.1&nbsp;&nbsp;</span>Apply</a></div><div class=\"lev2\"><a href=\"#Exercise\"><span class=\"toc-item-num\">1.10&nbsp;&nbsp;</span>Exercise</a></div><div class=\"lev2\"><a href=\"#References\"><span class=\"toc-item-num\">1.11&nbsp;&nbsp;</span>References</a></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Wrangling with Pandas\n",
    "\n",
    "Now that we have been exposed to the basic functionality of Pandas, lets explore some more advanced features that will be useful when addressing more complex data management tasks.\n",
    "\n",
    "As most statisticians/data analysts will admit, often the lion's share of the time spent implementing an analysis is devoted to preparing the data itself, rather than to coding or running a particular model that uses the data. This is where Pandas and  Python's standard library are beneficial, providing high-level, flexible, and efficient tools for manipulating your data as needed.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set_context('notebook')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Date/Time data handling\n",
    "\n",
    "Date and time data are inherently problematic. There are an unequal number of days in every month, an unequal number of days in a year (due to leap years), and time zones that vary over space. Yet information about time is essential in many analyses, particularly in the case of time series analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `datetime` built-in library handles temporal information down to the nanosecond."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datetime.datetime(2016, 10, 3, 13, 45, 55, 341774)"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "now = datetime.now()\n",
    "now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "now.day"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "now.weekday()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to `datetime` there are simpler objects for date and time information only, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from datetime import date, time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datetime.time(3, 24)"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time(3, 24)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datetime.date(1970, 9, 3)"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "date(1970, 9, 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having a custom data type for dates and times is convenient because we can perform operations on them easily. For example, we may want to calculate the difference between two times:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datetime.timedelta(17077, 49555, 341774)"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_age = now - datetime(1970, 1, 1)\n",
    "my_age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'datetime.timedelta'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "46.78630136986301"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(type(my_age))\n",
    "my_age.days/365"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this section, we will manipulate data collected from ocean-going vessels on the eastern seaboard. Vessel operations are monitored using the Automatic Identification System (AIS), a safety at sea navigation technology which vessels are required to maintain and that uses transponders to transmit very high frequency (VHF) radio signals containing static information including ship name, call sign, and country of origin, as well as dynamic information unique to a particular voyage such as vessel location, heading, and speed. \n",
    "\n",
    "The International Maritime Organizationâ€™s (IMO) International Convention for the Safety of Life at Sea requires functioning AIS capabilities on all vessels 300 gross tons or greater and the US Coast Guard requires AIS on nearly all vessels sailing in U.S. waters. The Coast Guard has established a national network of AIS receivers that provides coverage of nearly all U.S. waters. AIS signals are transmitted several times each minute and the network is capable of handling thousands of reports per minute and updates as often as every two seconds. Therefore, a typical voyage in our study might include the transmission of hundreds or thousands of AIS encoded signals. This provides a rich source of spatial data that includes both spatial and temporal information.\n",
    "\n",
    "For our purposes, we will use summarized data that describes the transit of a given vessel through a particular administrative area. The data includes the start and end time of the transit segment, as well as information about the speed of the vessel, how far it travelled, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mmsi</th>\n",
       "      <th>name</th>\n",
       "      <th>transit</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_length</th>\n",
       "      <th>avg_sog</th>\n",
       "      <th>min_sog</th>\n",
       "      <th>max_sog</th>\n",
       "      <th>pdgt10</th>\n",
       "      <th>st_time</th>\n",
       "      <th>end_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Us Govt Ves</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>13.2</td>\n",
       "      <td>9.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>96.5</td>\n",
       "      <td>2/10/09 16:03</td>\n",
       "      <td>2/10/09 16:27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Dredge Capt Frank</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>18.6</td>\n",
       "      <td>10.4</td>\n",
       "      <td>20.6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4/6/09 14:31</td>\n",
       "      <td>4/6/09 15:20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4.3</td>\n",
       "      <td>16.2</td>\n",
       "      <td>10.3</td>\n",
       "      <td>20.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4/6/09 14:36</td>\n",
       "      <td>4/6/09 14:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.5</td>\n",
       "      <td>16.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4/10/09 17:58</td>\n",
       "      <td>4/10/09 18:34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>Dredge Capt Frank</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.6</td>\n",
       "      <td>16.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4/10/09 17:59</td>\n",
       "      <td>4/10/09 18:35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mmsi               name  transit  segment  seg_length  avg_sog  min_sog  \\\n",
       "0     1        Us Govt Ves        1        1         5.1     13.2      9.2   \n",
       "1     1  Dredge Capt Frank        1        1        13.5     18.6     10.4   \n",
       "2     1      Us Gov Vessel        1        1         4.3     16.2     10.3   \n",
       "3     1      Us Gov Vessel        2        1         9.2     15.4     14.5   \n",
       "4     1  Dredge Capt Frank        2        1         9.2     15.4     14.6   \n",
       "\n",
       "   max_sog  pdgt10        st_time       end_time  \n",
       "0     14.5    96.5  2/10/09 16:03  2/10/09 16:27  \n",
       "1     20.6   100.0   4/6/09 14:31   4/6/09 15:20  \n",
       "2     20.5   100.0   4/6/09 14:36   4/6/09 14:55  \n",
       "3     16.1   100.0  4/10/09 17:58  4/10/09 18:34  \n",
       "4     16.2   100.0  4/10/09 17:59  4/10/09 18:35  "
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments = pd.read_csv(\"Data/AIS/transit_segments.csv\")\n",
    "segments.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example, we might be interested in the distribution of transit lengths, so we can plot them as a histogram:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x13534dba8>"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfsAAAFVCAYAAADonUZMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGpZJREFUeJzt3X+QXfV53/H3CmkRQVfyqF0pg+MfKm0fpjOtMjiDwNUP\nMFCBmpTmH8ow7uD8QAOjwUk7dseWK4chluU2CdXInqgtyEaBMBnD2GlrBoQSarwyTRCEFKshD7LJ\nQhumU0W70t4FvJJ2t3/cw3J1WbQXa8Xd8+X9mmH23ud898v3+OHqc865x0d9U1NTSJKkci3o9QIk\nSdK5ZdhLklQ4w16SpMIZ9pIkFc6wlySpcIa9JEmFWzjbgIhYCOwFPgqcAm4FJoD7gEngUGZuqcbe\nCmwGTgLbM/ORiFgMPACsAEaBWzLzaERcDuysxu7PzLvmdtckSRJ0d2a/CTgvM/8x8JvAl4G7ga2Z\nuQFYEBE3RMRK4A7gCuA6YEdELAJuB57PzPXA/cC2at7dwE2ZuQ5YExGr53LHJElSSzdh/yKwMCL6\ngGW0zsQvzczBavujwLXAZcCBzDyVmaPAYWA1sBZ4rG3s1RHRAPozc6iq7wOumYP9kSRJHWa9jA+M\nAauAvwT+FvALwLq27U1gKdAAjnf83rKOerOtNtoxx6p3v3xJkjSbbsL+XwGPZeYXIuKDwHeB/rbt\nDeAYrfBe2lEfqeqNjrHNGcYeO9Mipqampvr6+rpYriRJxZiT4Osm7IdpXbqHViAvBJ6LiA2Z+SRw\nPfAEcBDYHhH9wAXAJcAh4Cla3/s/U/0czMxmRIxHxCpgCNgI3HmmRfT19XHkSPPd7Z3mjYGBhv2r\nMftXX/au3gYGGrMP6kI3Yb8T+HpEfA9YBHwOeBa4t7oB7wXg4cyciohdwAFaRyJbM/NEROwG9kbE\nIDAO3FzNexvwIK37Bh7PzINzskeSJOk0fTX6W++mPDqtL88u6s3+1Ze9q7eBgcacXMb3oTqSJBXO\nsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpn2EuSVDjDXpKkwhn2kiQVzrCXJKlwhr0kSYUz7CVJ\nKpxhL0lS4Qx7SZIKZ9hLklQ4w16SpMIZ9pIkFc6wlySpcIa9JEmFM+wlSSqcYS9JUuEMe0mSCmfY\nS5JUOMNekqTC1TbsJyYm+NGPDjMxMdHrpUiSNK/VNuyHhl5i87Z7GRp6qddLkSRpXqtt2AMsXrK8\n10uQJGneq3XYS5Kk2Rn2kiQVzrCXJKlwhr0kSYVbONuAiLgF+BQwBVwArAbWATuBSeBQZm6pxt4K\nbAZOAtsz85GIWAw8AKwARoFbMvNoRFxezXES2J+Zd83xvkmSJLo4s8/MvZl5VWZ+AngW+DTwRWBr\nZm4AFkTEDRGxErgDuAK4DtgREYuA24HnM3M9cD+wrZp6N3BTZq4D1kTE6rneOUmS9C4u40fEzwH/\nIDPvBT6WmYPVpkeBa4HLgAOZeSozR4HDtK4CrAUeaxt7dUQ0gP7MHKrq+4BrznZnJEnS2816Gb/N\n54E7Z6g3gaVAAzjeVh8DlnXUm2210Y45Vs22gIGBxvTrkZElACxfvuS0uuYv+1Rv9q++7J26CvuI\nWAb8/cz8XlWabNvcAI7RCu+lHfWRqt7oGNucYeyx2dZx5Ehz+vXw8Nj0z/a65qeBgYZ9qjH7V1/2\nrt7m6kCt28v464E/bnv/XESsr15fDwwCB4G1EdFfHRxcAhwCngI2VWM3AYOZ2QTGI2JVRPQBG6s5\nJEnSHOv2Mn4A7Q+h/wxwT3UD3gvAw5k5FRG7gANAH60b+E5ExG5gb0QMAuPAzdUctwEP0jrgeDwz\nD5797kiSpE5dhX1m/nbH+8PAlTOM2wPs6ai9Adw4w9inad25L0mSziEfqiNJUuEMe0mSCmfYS5JU\nOMNekqTCGfaSJBXOsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpn2EuSVDjDXpKkwhn2kiQVzrCX\nJKlwhr0kSYUz7CVJKpxhL0lS4Qx7SZIKZ9hLklQ4w16SpMIZ9pIkFc6wlySpcIa9JEmFM+wlSSqc\nYS9JUuEMe0mSCmfYS5JUOMNekqTCGfaSJBXOsJckqXALuxkUEZ8D/hmwCPhd4HvAfcAkcCgzt1Tj\nbgU2AyeB7Zn5SEQsBh4AVgCjwC2ZeTQiLgd2VmP3Z+Zdc7ljkiSpZdYz+4jYAFyRmR8HrgQ+DNwN\nbM3MDcCCiLghIlYCdwBXANcBOyJiEXA78HxmrgfuB7ZVU+8GbsrMdcCaiFg9t7smSZKgu8v4G4FD\nEfGHwH8FvgNcmpmD1fZHgWuBy4ADmXkqM0eBw8BqYC3wWNvYqyOiAfRn5lBV3wdcMwf7I0mSOnRz\nGf9v0zqb/3ng79AK/PaDhCawFGgAx9vqY8CyjnqzrTbaMceqd798SZI0m27C/ijwQmaeAl6MiB8D\nP9O2vQEcoxXeSzvqI1W90TG2OcPYY7MtZGCgMf16ZGQJAMuXLzmtrvnLPtWb/asve6duwv4A8Gng\nP0TERcCFwB9HxIbMfBK4HngCOAhsj4h+4ALgEuAQ8BSwCXim+jmYmc2IGI+IVcAQra8K7pxtIUeO\nNKdfDw+PTf9sr2t+Ghho2Kcas3/1Ze/qba4O1GYN++qO+nUR8TTQR+uGuyHg3uoGvBeAhzNzKiJ2\n0To46KN1A9+JiNgN7I2IQWAcuLma+jbgQVpfCTyemQfnZI8kSdJpuvq/3mXm52YoXznDuD3Ano7a\nG8CNM4x9mtad+5Ik6RzyoTqSJBXOsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpn2EuSVDjDXpKk\nwhn2kiQVzrCXJKlwhr0kSYUz7CVJKpxhL0lS4Qx7SZIKZ9hLklQ4w16SpMIZ9pIkFc6wlySpcIa9\nJEmFq2XYT0xM8MorL/d6GZIk1UItw35o6CW27Xyo18uQJKkWahn2AOdfuKzXS5AkqRZqG/aSJKk7\nhr0kSYUz7CVJKpxhL0lS4Qx7SZIKZ9hLklQ4w16SpMIZ9pIkFW5hN4Mi4lngePX2r4AvA/cBk8Ch\nzNxSjbsV2AycBLZn5iMRsRh4AFgBjAK3ZObRiLgc2FmN3Z+Zd83ZXkmSpGmzntlHxPkAmfmJ6p9f\nAe4GtmbmBmBBRNwQESuBO4ArgOuAHRGxCLgdeD4z1wP3A9uqqXcDN2XmOmBNRKye652TJEndndmv\nBi6MiH3AecAXgEszc7Da/ijwT2id5R/IzFPAaEQcrn53LfDv2sb+24hoAP2ZOVTV9wHXAP/z7HdJ\nkiS16+Y7+9eB38rMjbTO0n8f6Gvb3gSWAg3eutQPMAYs66g322qjHXP4sHtJks6Bbs7sXwR+CJCZ\nhyPiKHBp2/YGcIxWeC/tqI9U9UbH2OYMY4/NtpCBgdY0IyNLpmvLly+Zrmt+s0/1Zv/qy96pm7D/\nZeAfAlsi4iJaIf14RGzIzCeB64EngIPA9ojoBy4ALgEOAU8Bm4Bnqp+DmdmMiPGIWAUMARuBO2db\nyJEjTQCGh8ema8PDY9N1zV8DAw37VGP2r77sXb3N1YFaN2G/B/hGRAzS+l7+U8BR4N7qBrwXgIcz\ncyoidgEHaF3m35qZJyJiN7C3+v1x4OZq3tuAB2l9lfB4Zh6ckz2SJEmnmTXsM/Mk8MkZNl05w9g9\ntA4O2mtvADfOMPZpWnfuS5Kkc8iH6kiSVDjDXpKkwhn2kiQVzrCXJKlwhr0kSYUz7CVJKpxhL0lS\n4Qx7SZIKZ9hLklQ4w16SpMIZ9pIkFc6wlySpcIa9JEmFM+wlSSqcYS9JUuEMe0mSCmfYS5JUOMNe\nkqTCGfaSJBXOsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpn2EuSVDjDXpKkwhn2kiQVzrCXJKlw\nhr0kSYUz7CVJKpxhL0lS4RZ2MygiVgDPANcAE8B9wCRwKDO3VGNuBTYDJ4HtmflIRCwGHgBWAKPA\nLZl5NCIuB3ZWY/dn5l1zuleSJGnarGf2EbEQ+I/A61XpbmBrZm4AFkTEDRGxErgDuAK4DtgREYuA\n24HnM3M9cD+wrZpjN3BTZq4D1kTE6rncKUmS9JZuLuP/Nq1wfhXoAy7NzMFq26PAtcBlwIHMPJWZ\no8BhYDWwFnisbezVEdEA+jNzqKrvo3XFQJIknQNnvIwfEZ8C/l9m7o+IrVW5/QChCSwFGsDxtvoY\nsKyj3myrjXbMsaqbxQ4MNAAYGVkyXVu+fMl0XfObfao3+1df9k6zfWf/S8BkRFxL60z994CBtu0N\n4Bit8F7aUR+p6o2Osc0Zxh7rZrFHjjQBGB4em64ND49N1zV/DQw07FON2b/6snf1NlcHame8jJ+Z\nGzLzqsy8Cvhz4F8Cj0bE+mrI9cAgcBBYGxH9EbEMuAQ4BDwFbKrGbgIGM7MJjEfEqojoAzZWc0iS\npHOgq7vxO3wGuKe6Ae8F4OHMnIqIXcABWt/rb83MExGxG9gbEYPAOHBzNcdtwIO0DjYez8yDZ7sj\nkiRpZl2HfWZ+ou3tlTNs3wPs6ai9Adw4w9inad25L0mSzjEfqiNJUuEMe0mSClfrsP/x2FFeeeXl\nXi9DkqR5rdZhL0mSZmfYS5JUOMNekqTCGfaSJBXOsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpn\n2EuSVDjDXpKkwhn2kiQVzrCXJKlwhr0kSYUz7CVJKpxhL0lS4Qx7SZIKZ9hLklQ4w16SpMIZ9pIk\nFc6wlySpcIa9JEmFM+wlSSqcYS9JUuEMe0mSCmfYS5JUOMNekqTCLZxtQEQsAO4BApgEbgPGgfuq\n94cyc0s19lZgM3AS2J6Zj0TEYuABYAUwCtySmUcj4nJgZzV2f2beNcf7JkmS6O7M/heAqcxcC2wD\nvgzcDWzNzA3Agoi4ISJWAncAVwDXATsiYhFwO/B8Zq4H7q/mANgN3JSZ64A1EbF6LndMkiS1zBr2\nmflfaJ2tA3wEGAEuzczBqvYocC1wGXAgM09l5ihwGFgNrAUeaxt7dUQ0gP7MHKrq+4Brzn53JElS\np66+s8/MyYi4D9gFPAj0tW1uAkuBBnC8rT4GLOuoN9tqox1zLHv3y5ckSbOZ9Tv7N2XmpyJiBXAQ\nuKBtUwM4Riu8l3bUR6p6o2Nsc4axx2Zbw8BAa5qRkSXTtWXLfmq6rvnNPtWb/asve6dubtD7JPAz\nmfkV4MfABPBMRGzIzCeB64EnaB0EbI+IfloHA5cAh4CngE3AM9XPwcxsRsR4RKwChoCNwJ2zreXI\nkSYAw8Nj07Xjx1+frmv+Ghho2Kcas3/1Ze/qba4O1Lo5s/8W8I2IeLIa/2ngL4F7qxvwXgAezsyp\niNgFHKB1mX9rZp6IiN3A3ogYpHUX/83VvLfR+kpgAfB4Zh6ckz2SJEmnmTXsM/N14F/MsOnKGcbu\nAfZ01N4Abpxh7NO07tyXJEnnkA/VkSSpcIa9JEmFM+wlSSqcYS9JUuEMe0mSCmfYS5JUOMNekqTC\nGfaSJBXOsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpn2EuSVDjDXpKkwhn2kiQVzrCXJKlwhr0k\nSYUz7CVJKpxhL0lS4Qx7SZIKZ9hLklQ4w16SpMLVOuynpiZ59dVXmZiY6PVSJEmat2od9ideH+Xr\nj/wvhoZe6vVSJEmat2od9gCLlyzv9RIkSZrXah/2kiTpzAx7SZIKZ9hLklQ4w16SpMItPNPGiFgI\nfB34KNAPbAf+ArgPmAQOZeaWauytwGbgJLA9Mx+JiMXAA8AKYBS4JTOPRsTlwM5q7P7MvGvud02S\nJMHsZ/afBP4mM9cD1wFfA+4GtmbmBmBBRNwQESuBO4ArqnE7ImIRcDvwfPX79wPbqnl3Azdl5jpg\nTUSsnusdkyRJLbOF/Td5K6DPA04Bl2bmYFV7FLgWuAw4kJmnMnMUOAysBtYCj7WNvToiGkB/Zg5V\n9X3ANXOwL5IkaQZnDPvMfD0zX6sC+iHgC0Bf25AmsBRoAMfb6mPAso56s6022jHHsrPYB0mSdAZn\n/M4eICI+BHwL+Fpm/kFE/Pu2zQ3gGK3wXtpRH6nqjY6xzRnGHutmsQMDralGRpacVl++fMn0Ns1f\n9qje7F992TvNdoPeSlqX2bdk5n+vys9FxPrM/B5wPfAEcBDYHhH9wAXAJcAh4ClgE/BM9XMwM5sR\nMR4Rq4AhYCNwZzeLPXKkCcDw8Nhp9eHhseltmp8GBhr2qMbsX33Zu3qbqwO12c7sPw98ANgWEV8E\npoBfA75a3YD3AvBwZk5FxC7gAK3L/Fsz80RE7Ab2RsQgMA7cXM17G/Agra8RHs/Mg3OyN5Ik6W3O\nGPaZ+evAr8+w6coZxu4B9nTU3gBunGHs07Tu3JckSeeYD9WRJKlwhr0kSYUz7CVJKpxhL0lS4Qx7\nSZIKZ9hLklQ4w16SpMIZ9pIkFc6wlySpcIa9JEmFM+wlSSqcYS9JUuEMe0mSClf7sJ+amuSVV15m\nYmKi10uRJGleqn3Yj782wpfu2c/Q0Eu9XookSfNS7cMeYPGS5b1egiRJ81YRYS9Jkt6ZYS9JUuEM\ne0mSCmfYS5JUOMNekqTCGfaSJBXOsJckqXCGvSRJhTPsJUkqnGEvSVLhDHtJkgpXu7CfmJjglVde\n7vUyJEmqjdqF/dDQS2zb+VCvlyFJUm3ULuwBzr9wWa+XIElSbdQy7CVJUvcWdjMoItYAX8nMqyLi\nYuA+YBI4lJlbqjG3ApuBk8D2zHwkIhYDDwArgFHglsw8GhGXAzursfsz86453i9JklSZ9cw+Ij4L\n3AOcX5XuBrZm5gZgQUTcEBErgTuAK4DrgB0RsQi4HXg+M9cD9wPbqjl2Azdl5jpgTUSsnsudkiRJ\nb+nmMv4PgV9se/+xzBysXj8KXAtcBhzIzFOZOQocBlYDa4HH2sZeHRENoD8zh6r6PuCas9oLSZL0\njma9jJ+Z346Ij7SV+tpeN4GlQAM43lYfA5Z11JtttdGOOVZ1s9iBgQYjI0tm3LZ8+RIGBhrdTKMe\nsT/1Zv/qy96pq+/sO0y2vW4Ax2iF99KO+khVb3SMbc4w9lg3/+IjR5oMD4/NuG14eIwjR5rdTKMe\nGBho2J8as3/1Ze/qba4O1H6Su/H/LCLWV6+vBwaBg8DaiOiPiGXAJcAh4ClgUzV2EzCYmU1gPCJW\nRUQfsLGaQ5IknQM/yZn9Z4B7qhvwXgAezsypiNgFHKB1mX9rZp6IiN3A3ogYBMaBm6s5bgMepHWw\n8XhmHjzbHZEkSTPrKuwz82Xg49Xrw8CVM4zZA+zpqL0B3DjD2Kdp3bkvSZLOMR+qI0lS4Qx7SZIK\nZ9hLklS4n+QGvZ74b489wYkfT/LBn17e66VIklQrtQn7p549DJOTfPCn1/R6KZIk1UqtLuNPTk7y\nyisv93oZkiTVSq3C/tjwEbbtfOht9amp1kHAxMRED1YlSdL8VquwBzj/wmVvq42/NsKX7tnP0NBL\nPViRJEnzW23CfnJykmMjR99x++Il3rgnSdJMahP2x4aP8Oc/fOewlyRJM6tN2MPMl/AlSdKZ1Srs\nJUnSu2fYS5JUOMNekqTCGfaSJBXOsJckqXCGvSRJhSsm7H1kriRJMysm7H1kriRJMysm7AHOv/AD\n/Mmf/A9efDF7vRRJkuaNosJ+/LURdn/z+/z1X//vXi9FkqR5o6iwBx+pK0lSp+LCXpIknc6wlySp\ncIa9JEmFM+wlSSqcYS9JUuEMe0mSCmfYS5JUuOLCfmpqkldffZUTJ07wox8d9ln5kqT3vYW9+hdH\nRB/wu8Bq4MfAr2bmWT/Y/sTro+z5zg8A2POdH7Bt80bWr7+K884772ynliSplnp5Zv/PgfMz8+PA\n54G752rivr4+dn/z+/T19fGb/3kf3/3uH/Hii+lZviTpfalnZ/bAWuAxgMz804j4ubmc/M3H5vb1\n9fEbu75F/08t5Qu/eg0XXfShM/7eeeedx8UX/12vBEiSitHLsF8KHG97fyoiFmTm5IyjT73B+GvH\nTyudeKMJ9M1aAxg7+n/43I6vn1Y7Of4ai86/8LRa/wUNvvSvb+LDH/5It/uhLoyMLGF4eKzXy9BP\nyP7Vl72bexdf/Pd6vYR3rZdhPwo02t6/c9AD/+mrO96e4JIkaVa9/M7++8AmgIi4HPhBD9ciSVKx\nenlm/23g2oj4fvX+l3q4FkmSitU3NTXV6zVIkqRzqLiH6kiSpNMZ9pIkFc6wlySpcIa9JEmF6+Xd\n+F05V8/Q19yLiGd560FJfwV8GbgPmAQOZeaWatytwGbgJLA9Mx9571crgIhYA3wlM6+KiIvpsl8R\nsRh4AFhB65kZt2Tm0V7sw/tZR/9+FvgO8GK1eXdmPmT/5p+IWAh8Hfgo0A9sB/6Cc/j5q8OZ/Tl7\nhr7mTkScD5CZn6j++RVavdqamRuABRFxQ0SsBO4ArgCuA3ZExKKeLfx9LCI+C9wDnF+V3k2/bgee\nz8z1wP3Atvd8B97nZujfx4DfafsMPmT/5q1PAn9T/e9/HfA1zvHnrw5hf9oz9IE5fYa+5sxq4MKI\n2BcRf1SdcVyamYPV9keBa4HLgAOZeSozR4HDwD/qzZLf934I/GLb+4912a/VtH0uq7HXvDdLVpu3\n9Q/4pxHxZETcExFLsH/z1Td5K6DPA07R/Z+XP1H/6hD2Mz5Dv1eL0Tt6HfitzNxI66jz9zn9Lylo\n0uplg9P7OQYse68Wqbdk5rdp/SHzpnfTr/b6m2P1Hpqhf38KfLY6M3wJ+A3e/uen/ZsHMvP1zHwt\nIhrAQ8AXOMefvzqE5rt6hr565kVaAU9mHgaOAivbtjeAY7T6uXSGunqv/XN1pn6NcPrn0h7OD3+Y\nmc+9+Rr4WVqBYP/moYj4EPAEsDcz/4Bz/PmrQ9j7DP16+GXgdwAi4iJa/4E+HhEbqu3XA4PAQWBt\nRPRHxDLgEuBQD9art/uziFhfvZ6tX09RfS6rn4Odk+k9t6/trwq/GngW+zcvVd/F7wP+TWburcrP\nncvP37y/Gx+foV8Xe4BvRMQgrSPUT9E6u7+3uqHkBeDhzJyKiF3AAVqXrbZm5okerVmn+wxwTzf9\niojdwN6q3+PAzT1btd50O/DViDgB/F9gc2aO2b956fPAB4BtEfFFYAr4NVr9OyefP5+NL0lS4epw\nGV+SJJ0Fw16SpMIZ9pIkFc6wlySpcIa9JEmFM+wlSSqcYS9JUuH+P/R0QzRY+MdXAAAAAElFTkSu\nQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1345ca6a0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "segments.seg_length.hist(bins=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Though most of the transits appear to be short, there are a few longer distances that make the plot difficult to read. This is where a transformation is useful:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1312bb128>"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfIAAAFVCAYAAAAUiG2GAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAHohJREFUeJzt3X9w3PV95/GnkLEMliziILhwIXXipp/M3WW4oTkCDTEk\nBxPgkuY608vlMpkjJLEHmuGa3pVc+KGGFlzalBIf5OIG2wRC2uawSZo2DHAkQGwgcyFpMtQN+dhI\nrIXlxnZt2bJlkPFq74/dlVdryZJWK3312X0+ZhhpP/vZ3fdHMnp9f3y+n29LoVBAkiSl6ZSsC5Ak\nSbUzyCVJSphBLklSwgxySZISZpBLkpQwg1ySpIQtmEqnEMK7gT+JMb4vhLAcuB8YAbbGGD9T6rMS\nWAW8DqyOMT4SQlgEfAM4CxgEro4x7gshXAisKfV9Isb4R3UelyRJTWHSPfIQwg3AOqCt1HQXcFOM\n8RLglBDCh0MIZwPXAxcBVwB3hBBOBa4DXogxrgAeBLpL77EW+GiM8b3Au0MI59VzUJIkNYupHFp/\nCfitise/HmPcUvr+UeBy4ALgmRjjsRjjILAdOA+4GHisou+/DyF0AAtjjLlS++PAZTMahSRJTWrS\nII8xfhs4VtHUUvH9IWAJ0AEcrGg/DHRWtR+qaBuseo/O6RYuSZKmeI68ykjF9x3AAYrBvKSqfaDU\n3lHV99A4fQ9M9qGFQqHQ0tIyWTdJkhrJpMFXS5D/fQhhRYxxM3Al8CTwPLA6hLAQOA14B7AVeA64\nCvhx6euWGOOhEMJwCOGtQA74AHDrpCNpaWHv3kM1lJuGrq4Ox5cwx5euRh4bOL7UdXV1TNqnliD/\nfWBdaTLbi8CmGGMhhHA38AzFrYebYoxHQwhrgQdCCFuAYeBjpfe4Fvgriof2/2+M8fka6pAkqem1\nJHT3s0Kjb3U5vnQ5vnQ18tjA8aWuq6tj0kPrLggjSVLCDHJJkhJmkEuSlDCDXJKkhBnkkiQlzCCX\nJClhBrkkSQkzyCVJSphBLklSwgxySZISZpBLkpQwg1ySpIQZ5JIkJcwglyQpYQa5NAP5fJ6enu3k\n8/msS5HUpAxyaQZyuV5Wda+np6cn61IkNSmDXJqhRe1Lsy5BUhMzyCVJSphBLklSwgxySZISZpBL\nNcrn8/T17ci6DElNziCXapTL9dK9ZmPWZUhqcga5NANtizuzLkFSkzPIJUlKmEEuSVLCDHJJkhJm\nkEuSlDCDXJKkhBnkkiQlzCCXJClhBrkkSQkzyCVJSphBLklSwgxySZISZpBLkpQwg1ySpIQZ5JIk\nJcwglyQpYQa5JEkJM8glSUqYQS5JUsIMckmSEmaQS3WWz+fp6dlOPp/PuhRJTcAgl+osl+tlVfd6\ncrnerEuR1AQMcmkWLGpfmnUJkpqEQS5JUsIMckmSEmaQS5KUMINckqSEGeSSJCXMIJckKWEGuSRJ\nCTPIJUlKmEEuSVLCFtTyohDCAuABYBlwDFgJ5IH7gRFga4zxM6W+K4FVwOvA6hjjIyGERcA3gLOA\nQeDqGOO+GY1EkqQmVOse+VVAa4zxPcBtwB8DdwE3xRgvAU4JIXw4hHA2cD1wEXAFcEcI4VTgOuCF\nGOMK4EGge4bjkCSpKdUa5NuABSGEFqCT4t72+THGLaXnHwUuBy4AnokxHosxDgLbgfOAi4HHKvpe\nVmMdkiQ1tZoOrQOHgbcCvwDeCHwIeG/F84eAJUAHcLDqdZ1V7eW+k+rq6qix3DQ4vrQMDLSPeVwe\nX7l96dL2hhpzI42lWiOPDRxfo6s1yH8PeCzGeHMI4V8CTwMLK57vAA5QPP+9pKp9oNTeUdV3Unv3\nHqqx3Pmvq6vD8SVm//7DYx6Xx1du37//cMOMuRF/f2WNPDZwfKmbykZKrYfW93N8j/oAxQ2Cn4YQ\nLim1XQlsAZ4HLg4hLAwhdALvALYCz1E8z07pa/mQvCRJmoZa98jXAPeFEDYDpwKfB34CrC9NZnsR\n2BRjLIQQ7gaeAVooToY7GkJYCzwQQtgCDAMfm+lAJElqRjUFeYxxCPjP4zx16Th9NwAbqtpeBT5S\ny2dL802hMMLLL7/MkiVn0dramnU5kpqMC8JIMzQ8NMANd36HXK4361IkNSGDXKqDRe1Lsy5BUpMy\nyCVJSphBLklSwgxySZISZpBLkpQwg1ySpIQZ5JIkJcwglyQpYQa5VIN8Pk9f346sy5Akg1yqRS7X\nS/eajVmXIUkGuVSrtsWdWZcgSQa5JEkpM8glSUqYQS5JUsIMckmSEmaQS5KUMINckqSEGeSSJCXM\nIJckKWEGuSRJCTPIJUlKmEEuSVLCDHJJkhJmkEuSlDCDXJKkhBnkkiQlbEHWBUgpyefz5HK95PMj\nWZciSYB75NK05HK9rOpeT3//K1mXIkmAQS5N26L2pVmXIEmjDHJJkhJmkEuSlDCDXJKkhBnkkiQl\nzCCXJClhBrkkSQkzyCVJSphBLklSwgxySZISZpBLkpQwg1ySpIQZ5JIkJcwglzKSz+fp6dlOPp/P\nuhRJCTPIpYyUb4may/VmXYqkhBnkUh0UCiP09e2Y9t61t0SVNFMGuVQHw0MD3L7uCfeuJc05g1yq\nE/euJWXBIJfqYHjoQNYlSGpSBrkkSQkzyCVJSphBLklSwgxyqU6OX4I2knUpkpqIQS7VSfkStP7+\nV7IuRVITWVDrC0MInwd+EzgV+AqwGbgfGAG2xhg/U+q3ElgFvA6sjjE+EkJYBHwDOAsYBK6OMe6b\nwTikecFL0CTNtZr2yEMIlwAXxRh/A7gUeAtwF3BTjPES4JQQwodDCGcD1wMXAVcAd4QQTgWuA16I\nMa4AHgS6ZzwSSZKaUK2H1j8AbA0h/A3wt8B3gfNjjFtKzz8KXA5cADwTYzwWYxwEtgPnARcDj1X0\nvazGOiRJamq1Hlo/k+Je+AeBt1EM88qNgkPAEqADOFjRfhjorGov95UkSdNUa5DvA16MMR4DtoUQ\nXgPeXPF8B3CA4vnvJVXtA6X2jqq+k+rq6pi8U8Ic3/w3MNAOQGfn6RP26ew8nUJhhIMH97J06Xm0\ntrae9L2WLm1P4meTQo21auSxgeNrdLUG+TPAfwO+FEI4B1gMfD+EcEmM8QfAlcCTwPPA6hDCQuA0\n4B3AVuA54Crgx6WvW078iBPt3XuoxnLnv66uDseXgP37DwNw8OCRCfscPHiE4aEBbrjzO9zb2cXy\n5W8/6Xvt33943v9sGuX3N55GHhs4vtRNZSOlpiAvzTx/bwjhR0ALxclrOWB9aTLbi8CmGGMhhHA3\nxeBvoTgZ7mgIYS3wQAhhCzAMfKyWOqT5zBnskuZCzZefxRg/P07zpeP02wBsqGp7FfhIrZ8tSZKK\nXBBGmqbXDu9j165dWZchSYBBLklS0gxySZISZpBLkpQwg1ySpIQZ5JIkJcwglyQpYQa5JEkJM8gl\nSUqYQS7Nor6+HfT0bM+6DEkNzCCX6qhQGHHVN0lzyiCX6mh4aIC1Dz2bdRmSmohBLk1ToTDCnj27\nJ3y+bXHnHFYjqdkZ5NI0HT0yyMNPv5R1GZIEGORSTdzrljRfGOSSJCXMIJckKWEGuZShQmGEvr4d\n5PP5rEuRlCiDXMrQ8NAAt697glyuN+tSJCXKIJemKJ/P09e3o+7vu6h9ad3fU1LzMMilKcrleule\nszHrMiRpDINcmgYvO5M03xjkkiQlzCCXJClhBrk0S8p3QsvnR7IuRVIDM8ilWVK+E1p//ytZlyKp\ngRnk0ixycpyk2WaQS5KUMINckqSEGeSSJCXMIJckKWEGuSRJCTPIJUlKmEEuSVLCDHJJkhJmkEuS\nlDCDXJKkhBnkkiQlzCCXJClhBrkkSQkzyCVJSphBLklSwgxySZISZpBLkpQwg1ySpIQZ5JIkJcwg\nlyQpYQa5JEkJM8glSUqYQS5JUsIMckmSEmaQS5KUMINcykA+n6evb0fWZUhqAAtm8uIQwlnAj4HL\ngDxwPzACbI0xfqbUZyWwCngdWB1jfCSEsAj4BnAWMAhcHWPcN5NapJTkcr10r9lI2+JOoBjsPT3b\nWbbsbbS2tmZcnaSU1LxHHkJYAPwFcKTUdBdwU4zxEuCUEMKHQwhnA9cDFwFXAHeEEE4FrgNeiDGu\nAB4EumcwBilJ5RAH6O/fyaru9eRyvRlWJClFMzm0fiewFtgFtADnxxi3lJ57FLgcuAB4JsZ4LMY4\nCGwHzgMuBh6r6HvZDOqQGsKi9qVZlyApQTUFeQjhE8CeGOMTFEO8+r0OAUuADuBgRfthoLOqvdxX\nkiRNU63nyK8BRkIIl1Pcw/460FXxfAdwgOL57yVV7QOl9o6qvpPq6uqYvFPCHN/8NjDQXtPr2tvb\nGBj4J5YvXz56/rv6vTo7Twdg6dL2eftzmq911UMjjw0cX6OrKchL58EBCCE8CVwL/FkIYUWMcTNw\nJfAk8DywOoSwEDgNeAewFXgOuIriRLmrgC1Mwd69h2opNwldXR2Ob57bv/9wTa/btu1lPv+l73Lv\nbZ9m+fK3n/BehcIIL7740mj7fPw5NcLvbyKNPDZwfKmbykZKPS8/+33gj0IIzwKnAptijLuBu4Fn\ngO9RnAx3lOK59X8TQtgCfBr4wzrWIc07Jzv/PTw0wNqHnp3DaiQ1khldfgYQY3x/xcNLx3l+A7Ch\nqu1V4CMz/WypUVTOYJek6XBBGEmSEmaQS5KUMINckqSEGeSSJCXMIJckKWEGuSRJCTPIpVlUKIyw\nZ8/urMuQ1MAMcmkWHT0yyMNPv5R1GZIamEEuzTIXe5E0mwxySZISZpBLkpQwg1yaB4aHpnQnX0k6\ngUEuSVLCDHJJkhJmkEuSlLAZ349c0tTl83k2b36KfH4k61IkNQj3yKU5lMv10r1mI7t3/zLrUiQ1\nCINcmmMuECOpnjy0Ls2RfD5Pf//OrMuQ1GDcI5fmSH//TrrXbMy6DEkNxiCX5pCH1SXVm0EuSVLC\nDHJJkhJmkEuSlDCDXJKkhBnkkiQlzCCXJClhBrkkSQkzyCVJSphBLklSwgxySZISZpBLkpQwg1ya\np3p6ttPTsz3rMiTNcwa5JEkJM8glSUqYQS5JUsIMcmkS+Xyenp7t5PMjs/o5hcIIfX07yOfzs/o5\nkhqLQS5NIpfrZVX3evr7X5nVzxkeGuD2dU+MTnLL50cqNiIMd0njM8ilKVjUvnTOPqe/f+fohkP5\n+1yud04+X1J6DHJpjhUKI+zZs/ukfSo3HOZqI0JSmgxyaY4dPTLIw0+/lHUZkhqEQS5NQaEwwq5d\nu+r2fm2LO+v2XpKam0EuTcHw0ABrH3q25tfXe0NAksoMcmmKZrIXPZMNAS9Lk3QyBrk0R2rdEChf\nlubMdUnjMcilBDhzXdJEDHJJkhJmkEuSlDCDXJKkhBnk0jxUvlxttm/UIil9Brk0D5UvV9u9+5dj\n2r2JiqRqBrk0T413uVr5TmxeiiapzCCXTiKfz9PXtyPrMsbwUjRJlQxy6SRyuV6612ycs89zKVdJ\n07WglheFEBYA9wHLgIXAauDnwP3ACLA1xviZUt+VwCrgdWB1jPGREMIi4BvAWcAgcHWMcd+MRiLN\nkrm8wUnx3HiOtsWdU7rdqSTVukf+ceCfY4wrgCuALwN3ATfFGC8BTgkhfDiEcDZwPXBRqd8dIYRT\ngeuAF0qvfxDonuE4pIZR3nDwdqeSpqLWIH+I4+HbChwDzo8xbim1PQpcDlwAPBNjPBZjHAS2A+cB\nFwOPVfS9rMY6pIbm7U4lTaamQ+sxxiMAIYQOYCNwM3BnRZdDwBKgAzhY0X4Y6KxqL/edVFdXRy3l\nJsPxzT8DA+1ZlzBq6dL2Md/P9c8zxd/fVDXy2MDxNbqaghwghHAu8C3gyzHGb4YQvljxdAdwgOL5\n7yVV7QOl9o6qvpPau/dQreXOe11dHY5vHtq//3DWJYzau/cg/f07ee3wPn72s3/kDW9405x9dqq/\nv6lo5LGB40vdVDZSajq0Xjr3/TjwuRjjA6Xmn4YQVpS+vxLYAjwPXBxCWBhC6ATeAWwFngOuKvW9\nqtRX0kn09++c0xn0ktJQ6x75jcAZQHcI4Q+AAvC7wD2lyWwvAptijIUQwt3AM0ALxclwR0MIa4EH\nQghbgGHgYzMdiNQMyrPZi8u35mltbc26JEkZq/Uc+WeBz47z1KXj9N0AbKhqexX4SC2fLTW7o0cG\nue+Rf+TCC3tZvvztWZcjKWMuCCMlaFH7UtddlwQY5FKSCoURnn/+R667Lskgl1JUvjua665LMsil\nBIy3BruLxUgCg1ya0Hy681l5D1ySqhnk0gTm+s5nk3EPXNJ4DHLpJAxPSfOdQS5JUsIMckmSEmaQ\nS5KUMINckqSEGeSSJCWs5vuRS40qn8+Ty/WSz49kXYokTco9cqlKLtfLqu719Pe/knUpkjQpg1wa\nh2uYS0qFQS5JUsIMckmSEmaQS5KUMGetSyXOVpeUIvfIpZLxZqsPDx3IsKLJFQoj5HK9bNsWyefz\nWZcjKQMGuVQhtdnqw0MDfOHub3HdrfeRy/VmXY6kDBjkUuLaFnfOaAMkn8/T07PdPXopUQa51OTK\npxTco5fS5GQ3qUlVTu5L7ZSCpOPcI1fTO35oublmq7sUrdQYDHI1vfECrVAYYdeuXRlWNTfcE5fS\n56F1iRMDbXhogLUP5bIpRpKmwT1yaQJtizuzLmHKCoUR+vp2OPNcakIGudQAhocGuH3dE848l5qQ\nQS5VSPncuOe7peZkkKvp9PRsp6dn+7jPFc+NPzvHFc1ceSnZ6S7ukvKGi6Qig1yqktK58Wr9/Tun\ntbhLqhsuko4zyCUaY8+0PIbpHmJPecNFkkEuAY2xZ1oew1RmsOfzefr6dsxhdZJmi0EuUTzH3Ah7\npm2LOxkeGuC2ex9n8+anOHr0KD0920e/lsM9l+ule83GjKuVVA8GudSAWlpauH3dEzz77GZWda8f\n/Vp57rwRNlwkGeRqYM1+e87yufK2xWfUdO5cUhoMcjWsqdyes5HPFb96aC8/+9lPG+L8v6SJuda6\nGs5Et+esbIcC27b9gr6+Pm695+Hsip1FR48M8vDTh2hb3DnmMHojb7xIzcggV0PJ5/Ns3vwUt697\ngltWXs5rh/eN7pHn8yNc+4X1XHPVvwbgq5t+CJQniB3IrObZVBnghcIIuVwvzz33LF/d9EPPkUsN\nwiBXQynPxu48+21AMbxeeOEFVq//Hp/64DtpaWnhngcfZ+FpS5ouyIaHBvjC3d/i6KuDdJx57pjn\nXju8j76+HSxf/vaMqpNUK8+Rq2GUDxlXBnTx8PJLtLS0jJ4nXnhaR9OFeFnb4k4WntaRdRmS6sgg\nV0MoH1If79rocmg3a3hLamwGuZJXGeKGtaRmk2yQj3eNcLNfN9xoqn+f5ceVq5Vt2/YLnnrq+4a4\npKaVbJCPd43wVK4b1vxysg2ynp6Xxvw+c7leVt5yLxs3fnN0tbJP3vAlbr3nYUNcUtNKNsiBcVeq\nqm6rvvd0ZXC4Bz97pvqzLW98lX9P5Wu9V3Wvp7//FRa1L614r5HRSWvHVy3rHDfEG+FuZpI0FUkH\neS0q99rdg589J/vZVm9cLWpfesJ9tCs3yMrP9fe/Akxt0pqrmU1PoTDCzp072bYtumErJabpghzG\nhkT1yl8nOyfrH7jpme7a3ifrX8s64R5un7qjRwb56qYfcu0X1rN581OT3gLV/x+k+SOZIN+3bx9H\njx6d1c+o3ossP66+c5QBP77Kn8tMlwCtPjTuofLZ17a4c/SuaeWjJtX/xstXCKy85d5JA1/S3Egm\nyFfdcCcPf+e7s/451Xt+5ceV7RMFfFnlHkvlud/p7O3Xc2Mhn8+zbdsv2LYtnjDju7JtstpOtidW\n+Qd+48Zv0r1m4+iSoOXPOP55r9PXt2P0fcohXe7/8ssvn3Bo3EPlc6dt8Rk8//yPxkws3LZt2+gE\nxP95xwZaWlq47d7Hefrp7/Hiiz/3kLyUoWSWaD19yZkUCllXcdx4AV953ndV93ruve3To4/L4b/2\n1mtobW0lnx/hulvv45aVl3P7uifY9L8/yxve8KYxN/aofL78unPP/ZXShkMLy5a9ddLvAfr7Xyld\nnnUGt6y8nO41G7nuI+9h7UPPjrbddu/jfOqD7+T889/FL3+5a/Txhu/+Azd/+jLOOedc+vt38scb\nnhh93NraOvpZ5ZuPtC3uLL1vJ8NDA9z4p/fT/sY389n/8u/44obvs/D0JXzogrP4ux/t4eaRPIVC\nSymkc7Qt7uQLd38LOD6JrXwHr3IbwPDQAdoWv2HMHnqx7YxZ+E03n/LvAwqsfehZlpy1jOeee441\nf/08n/rgO0dXhmtpaRn9fS08fQndqz7AihXvo7W1NbvipSaUTJCnZqIZ9f39O0dv6FG9MdDTs52+\nvh3jPl/5uupQnux7KIZg5Uzv6rbybPDrYDSIy1/Lf6zLr7nxT+8vrVV+/LOKCmPeH4rLoba0tPDF\nDd8fbf/rR/+ejjPPHRPa470Wxt7Bq1Jl+Kv+yjeSKW+QlX9/1UdFKjeubrv3cW4eyXPOOcfXcW9t\nbWX58l8tbbwWN1KXLXvbmLAvH+mBltG+lSZ6naSizII8hNACfAU4D3gN+HSMsSmmj082cWui5yuD\neDrfT1X1UqYTLW1aXqu8esPgZHcQqw738d53Kq+dSrvqr/LfwkS/58o9dICjrw7S/sY3T3hEp6y/\nfye33vMwC09fcsJz5efLrzv33GXjhr3UzLLcI/+PQFuM8TdCCO8G7iq1SUrU2I2rwgnhXn2E52R9\nj746yMLTlox5XaEwwrX/6T28610XjPar3OuXmlGWQX4x8BhAjPH/hRDelWEtkmbJiUdOCifMZyjv\n5VdvCFQ+Lh8N+OqmH47eSx6YcE9+qnbvXsz+/UOT9ivPCXnllR1j5qosX/6rANM+PVC50NFUNkbq\neYphstMZ9eApkbmTZZAvAQ5WPD4WQjglxjgyXuf80F4OHjhldEJZX98OXju8f8xlThO1VaruU/l9\n9XPlx7t27Zpye1l1HeW28muqv7788sscPHhkwucrvw4PHQRapvw9cMJze/bsHqc/Fe0TO/rqIYoT\n6ca+9nj7eH2n39bsr5+PNWX9+qOvHhrdSx9tOzLI/7j1Hk5tWzym/fXhoSm1TbXvwtM6+O33BzY9\nGfmdj65gzdceYeFpHdz+3z8KwOdWrxt9/Ja3/MqY9+rr28Etd31zzHN9fTv43Op1o+893uuq3+Nz\nq9fx2Wv+AxdeeNGE/aoNDLSzf//hcd9rKp9bq/HGPBvGG99cWL787XP+mRNpKWQ0FTyE8OfAD2OM\nm0qP+2KMb8mkGEmSEpXldeTPAlcBhBAuBP4hw1okSUpSlofWvw1cHkIoX89yTYa1SJKUpMwOrUuS\npJlLZolWSZJ0IoNckqSEGeSSJCXMIJckKWHz/qYpzbAme2mJ2j+JMb4v61rqKYSwALgPWAYsBFbH\nGP8u06LqKIRwCrAOCMAIcG2M8efZVlV/IYSzgB8Dl8UYt2VdTz2FEH7C8YWpXo4xfirLeuothPB5\n4DeBU4GvxBi/lnFJdRNCuBr4BMU7Lp1GMSP+RYxxMMu66qH0t/MBin87jwErT/b/Xgp75KNrsgM3\nUlyTvWGEEG6gGAZtWdcyCz4O/HOMcQVwJfDljOuptw8BhRjjxUA38McZ11N3pT8ofwEcybqWegsh\ntAHEGN9f+q/RQvwS4KLS385LgdrWsJ2nYowPxBjfF2N8P/AT4PpGCPGSq4DWGON7gNuY5G9LCkE+\nZk12oNHWZH8J+K2si5glD1EMOCj+W3s9w1rqLsb4HWBV6eEyYCC7ambNncBaYNdkHRN0HrA4hPB4\nCOF7pSNjjeQDwNYQwt8Afwt8N+N6ZkXpPh3/Ksa4Ieta6mgbsKB0RLoTOHqyzikE+bhrsmdVTL3F\nGL9N8dBJw4kxHokxDoUQOoCNwM1Z11RvMcaREML9wP8C/jLjcuoqhPAJYE+M8QnGWzQ9fUeAP4sx\nfgC4DvjLRvrbApwJ/Drw2xTH91fZljNrbgT+MOsi6uww8FbgF8BXgbtP1jmFf7SDQEfF4wlvrKL5\nJ4RwLvAk8ECM8f9kXc9siDF+Avg1YH0I4bSMy6mnayiuvvgU8G+Br5fOlzeKbZQ2vmKM24F9wJsy\nrai+9gGPxxiPlc6vvhZCODProuophNAJ/FqM8QdZ11Jnvwc8FmMMFI8cfT2EsHCizikEebOsyd5w\nezwhhLOBx4HPxRgfyLqeegshfLw0mQiKEzHzFCe9NYQY4yWlc5DvA34G/NcY456s66qjTwJ/DhBC\nOIfiDsM/ZVpRfT0DXAGj4zudYrg3khXA97MuYhbs5/iR6AMUJ6ZPeC/YeT9rneZZk70R18q9ETgD\n6A4h/AHFMV4ZYxzOtqy6+RbwtRDCDyj+v/S7DTS2ao3473MDxd/fFoobYJ9spKN9McZHQgjvDSH8\niOKOwu/EGBvt9xiAhrqKqWQNcF8IYTPFKw5ujDG+OlFn11qXJClhKRxalyRJEzDIJUlKmEEuSVLC\nDHJJkhJmkEuSlDCDXJKkhBnkkiQl7P8De+IT7lX9RdkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x13128aa20>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "segments.seg_length.apply(np.log).hist(bins=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that although there are date/time fields in the dataset, they are not in any specialized format, such as `datetime`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.st_time.dtype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our first order of business will be to convert these data to `datetime`. The `strptime` method parses a string representation of a date and/or time field, according to the expected format of this information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datetime.datetime(2009, 2, 10, 16, 3)"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datetime.strptime(segments.st_time.ix[0], '%m/%d/%y %H:%M')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `dateutil` package includes a parser that attempts to detect the format of the date strings, and convert them automatically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from dateutil.parser import parse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datetime.datetime(2009, 2, 10, 16, 3)"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse(segments.st_time.ix[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can convert all the dates in a particular column by using the `apply` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        2009-02-10 16:03:00\n",
       "1        2009-04-06 14:31:00\n",
       "2        2009-04-06 14:36:00\n",
       "3        2009-04-10 17:58:00\n",
       "4        2009-04-10 17:59:00\n",
       "5        2010-03-20 16:06:00\n",
       "6        2010-03-20 18:05:00\n",
       "7        2011-05-04 11:28:00\n",
       "8        2010-06-05 11:23:00\n",
       "9        2010-06-08 11:03:00\n",
       "10       2010-06-14 17:07:00\n",
       "11       2010-06-17 16:35:00\n",
       "12       2010-07-11 11:51:00\n",
       "13       2010-07-12 15:33:00\n",
       "14       2010-07-20 19:23:00\n",
       "15       2010-09-10 14:00:00\n",
       "16       2011-03-14 16:13:00\n",
       "17       2011-03-18 11:18:00\n",
       "18       2011-04-25 16:37:00\n",
       "19       2011-05-14 15:51:00\n",
       "20       2011-05-19 12:34:00\n",
       "21       2011-05-23 21:29:00\n",
       "22       2011-05-25 11:46:00\n",
       "23       2011-06-01 14:32:00\n",
       "24       2011-06-07 14:43:00\n",
       "25       2011-08-08 15:42:00\n",
       "26       2011-09-06 20:59:00\n",
       "27       2011-09-08 08:22:00\n",
       "28       2011-09-12 14:56:00\n",
       "29       2011-10-25 12:52:00\n",
       "                 ...        \n",
       "262496   2009-11-22 02:13:00\n",
       "262497   2009-11-22 02:15:00\n",
       "262498   2009-12-02 01:46:00\n",
       "262499   2009-12-02 03:12:00\n",
       "262500   2009-12-06 16:09:00\n",
       "262501   2009-12-06 16:21:00\n",
       "262502   2009-12-11 23:49:00\n",
       "262503   2009-12-15 22:23:00\n",
       "262504   2009-12-15 22:51:00\n",
       "262505   2009-12-31 10:17:00\n",
       "262506   2010-01-01 11:22:00\n",
       "262507   2010-01-13 17:40:00\n",
       "262508   2010-01-13 18:07:00\n",
       "262509   2010-01-24 05:53:00\n",
       "262510   2010-03-20 05:14:00\n",
       "262511   2010-04-14 15:09:00\n",
       "262512   2010-04-19 00:20:00\n",
       "262513   2010-05-13 14:13:00\n",
       "262514   2010-05-19 19:46:00\n",
       "262515   2010-05-31 14:27:00\n",
       "262516   2010-06-05 05:25:00\n",
       "262517   2010-06-27 02:35:00\n",
       "262518   2010-07-01 03:49:00\n",
       "262519   2010-07-02 03:30:00\n",
       "262520   2010-06-13 10:32:00\n",
       "262521   2010-06-15 12:49:00\n",
       "262522   2010-06-15 21:32:00\n",
       "262523   2010-06-17 19:16:00\n",
       "262524   2010-06-18 02:52:00\n",
       "262525   2010-06-18 10:19:00\n",
       "Name: st_time, dtype: datetime64[ns]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.st_time.apply(lambda d: datetime.strptime(d, '%m/%d/%y %H:%M'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a convenience, Pandas has a `to_datetime` method that will parse and convert an entire Series of formatted strings into `datetime` objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   2009-02-10 16:03:00\n",
       "1   2009-04-06 14:31:00\n",
       "2   2009-04-06 14:36:00\n",
       "3   2009-04-10 17:58:00\n",
       "4   2009-04-10 17:59:00\n",
       "5   2010-03-20 16:06:00\n",
       "6   2010-03-20 18:05:00\n",
       "7   2011-05-04 11:28:00\n",
       "8   2010-06-05 11:23:00\n",
       "9   2010-06-08 11:03:00\n",
       "Name: st_time, dtype: datetime64[ns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.to_datetime(segments.st_time[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas also has a custom NA value for missing datetime objects, `NaT`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DatetimeIndex(['NaT'], dtype='datetime64[ns]', freq=None)"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.to_datetime([None])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also, if `to_datetime()` has problems parsing any particular date/time format, you can pass the spec in using the `format=` argument."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `read_*` functions now have an optional `parse_dates` argument that try to convert any columns passed to it into `datetime` format upon import:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "segments = pd.read_csv(\"Data/AIS/transit_segments.csv\", parse_dates=['st_time', 'end_time'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mmsi                   int64\n",
       "name                  object\n",
       "transit                int64\n",
       "segment                int64\n",
       "seg_length           float64\n",
       "avg_sog              float64\n",
       "min_sog              float64\n",
       "max_sog              float64\n",
       "pdgt10               float64\n",
       "st_time       datetime64[ns]\n",
       "end_time      datetime64[ns]\n",
       "dtype: object"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns of the `datetime` type have an **accessor** to easily extract properties of the data type. This will return a `Series`, with the same row index as the `DataFrame`. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    2\n",
       "1    4\n",
       "2    4\n",
       "3    4\n",
       "4    4\n",
       "Name: st_time, dtype: int64"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.st_time.dt.month.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    16\n",
       "1    14\n",
       "2    14\n",
       "3    17\n",
       "4    17\n",
       "Name: st_time, dtype: int64"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.st_time.dt.hour.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This can be used to easily filter rows by particular temporal attributes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mmsi</th>\n",
       "      <th>name</th>\n",
       "      <th>transit</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_length</th>\n",
       "      <th>avg_sog</th>\n",
       "      <th>min_sog</th>\n",
       "      <th>max_sog</th>\n",
       "      <th>pdgt10</th>\n",
       "      <th>st_time</th>\n",
       "      <th>end_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Us Govt Ves</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>13.2</td>\n",
       "      <td>9.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>96.5</td>\n",
       "      <td>2009-02-10 16:03:00</td>\n",
       "      <td>2009-02-10 16:27:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>3011</td>\n",
       "      <td>Charleston</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>18.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2010-02-07 07:26:00</td>\n",
       "      <td>2010-02-09 19:27:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>3011</td>\n",
       "      <td>Charleston</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>19.2</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2010-02-11 16:56:00</td>\n",
       "      <td>2010-02-14 14:31:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>3011</td>\n",
       "      <td>Charleston</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>24.7</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2010-02-19 11:53:00</td>\n",
       "      <td>2010-02-22 16:50:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>3011</td>\n",
       "      <td>Charleston</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>40.1</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>2010-02-23 15:15:00</td>\n",
       "      <td>2010-03-02 14:25:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    mmsi         name  transit  segment  seg_length  avg_sog  min_sog  \\\n",
       "0      1  Us Govt Ves        1        1         5.1     13.2      9.2   \n",
       "78  3011   Charleston       16        1        18.9      0.3      0.0   \n",
       "79  3011   Charleston       17        1        19.2      0.3      0.0   \n",
       "80  3011   Charleston       18        1        24.7      0.3      0.0   \n",
       "81  3011   Charleston       19        1        40.1      0.4      0.0   \n",
       "\n",
       "    max_sog  pdgt10             st_time            end_time  \n",
       "0      14.5    96.5 2009-02-10 16:03:00 2009-02-10 16:27:00  \n",
       "78     16.1     0.5 2010-02-07 07:26:00 2010-02-09 19:27:00  \n",
       "79      6.4     0.0 2010-02-11 16:56:00 2010-02-14 14:31:00  \n",
       "80      5.7     0.0 2010-02-19 11:53:00 2010-02-22 16:50:00  \n",
       "81     16.3     0.1 2010-02-23 15:15:00 2010-03-02 14:25:00  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments[segments.st_time.dt.month==2].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition, time zone information can be applied:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   2009-02-10 16:03:00+00:00\n",
       "1   2009-04-06 14:31:00+00:00\n",
       "2   2009-04-06 14:36:00+00:00\n",
       "3   2009-04-10 17:58:00+00:00\n",
       "4   2009-04-10 17:59:00+00:00\n",
       "Name: st_time, dtype: datetime64[ns, UTC]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.st_time.dt.tz_localize('UTC').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   2009-02-10 11:03:00-05:00\n",
       "1   2009-04-06 10:31:00-04:00\n",
       "2   2009-04-06 10:36:00-04:00\n",
       "3   2009-04-10 13:58:00-04:00\n",
       "4   2009-04-10 13:59:00-04:00\n",
       "Name: st_time, dtype: datetime64[ns, US/Eastern]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.st_time.dt.tz_localize('UTC').dt.tz_convert('US/Eastern').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging and joining DataFrame objects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the vessel transit information as we need it, we may want a little more information regarding the vessels themselves. In the `data/AIS` folder there is a second table that contains information about each of the ships that traveled the segments in the `segments` table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_names</th>\n",
       "      <th>names</th>\n",
       "      <th>sov</th>\n",
       "      <th>flag</th>\n",
       "      <th>flag_type</th>\n",
       "      <th>num_loas</th>\n",
       "      <th>loa</th>\n",
       "      <th>max_loa</th>\n",
       "      <th>num_types</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mmsi</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>000000009/Raven/Shearwater</td>\n",
       "      <td>N</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>50.0/62.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Pleasure/Tug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>208.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>2</td>\n",
       "      <td>Mcfaul/Sarah Bell</td>\n",
       "      <td>N</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>155.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>3</td>\n",
       "      <td>Ron G/Us Navy Warship 103/Us Warship 103</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>26.0/155.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Tanker/Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      num_names                                              names sov  \\\n",
       "mmsi                                                                     \n",
       "1             8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y   \n",
       "9             3                         000000009/Raven/Shearwater   N   \n",
       "21            1                                      Us Gov Vessel   Y   \n",
       "74            2                                  Mcfaul/Sarah Bell   N   \n",
       "103           3           Ron G/Us Navy Warship 103/Us Warship 103   Y   \n",
       "\n",
       "         flag flag_type  num_loas                                    loa  \\\n",
       "mmsi                                                                       \n",
       "1     Unknown   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0   \n",
       "9     Unknown   Unknown         2                              50.0/62.0   \n",
       "21    Unknown   Unknown         1                                  208.0   \n",
       "74    Unknown   Unknown         1                                  155.0   \n",
       "103   Unknown   Unknown         2                             26.0/155.0   \n",
       "\n",
       "      max_loa  num_types                             type  \n",
       "mmsi                                                       \n",
       "1       156.0          4  Dredging/MilOps/Reserved/Towing  \n",
       "9        62.0          2                     Pleasure/Tug  \n",
       "21      208.0          1                          Unknown  \n",
       "74      155.0          1                          Unknown  \n",
       "103     155.0          2                   Tanker/Unknown  "
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vessels = pd.read_csv(\"Data/AIS/vessel_information.csv\", index_col='mmsi')\n",
    "vessels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Unknown',\n",
       " 'Other',\n",
       " 'Tug',\n",
       " 'Towing',\n",
       " 'Pleasure',\n",
       " 'Cargo',\n",
       " 'WIG',\n",
       " 'Fishing',\n",
       " 'BigTow',\n",
       " 'MilOps',\n",
       " 'Tanker',\n",
       " 'Passenger',\n",
       " 'SAR',\n",
       " 'Sailing',\n",
       " 'Reserved',\n",
       " 'Law',\n",
       " 'Dredging',\n",
       " 'AntiPol',\n",
       " 'Pilot',\n",
       " 'HSC',\n",
       " 'Diving',\n",
       " 'Resol-18',\n",
       " 'Tender',\n",
       " 'Spare',\n",
       " 'Medical']"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[v for v in vessels.type.unique() if v.find('/')==-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cargo                                                                                5622\n",
       "Tanker                                                                               2440\n",
       "Pleasure                                                                              601\n",
       "Tug                                                                                   221\n",
       "Sailing                                                                               205\n",
       "Fishing                                                                               200\n",
       "Other                                                                                 178\n",
       "Passenger                                                                             150\n",
       "Towing                                                                                117\n",
       "Unknown                                                                               106\n",
       "MilOps                                                                                 82\n",
       "Cargo/Unknown                                                                          79\n",
       "Cargo/Other                                                                            63\n",
       "Cargo/Tanker                                                                           42\n",
       "Towing/Tug                                                                             30\n",
       "Tanker/Unknown                                                                         30\n",
       "BigTow                                                                                 29\n",
       "Fishing/Pleasure                                                                       24\n",
       "Dredging                                                                               24\n",
       "Reserved                                                                               20\n",
       "Cargo/Reserved                                                                         19\n",
       "BigTow/Towing                                                                          19\n",
       "Other/Tanker                                                                           16\n",
       "BigTow/Tug                                                                             15\n",
       "Cargo/MilOps                                                                           15\n",
       "Tug/Unknown                                                                            15\n",
       "HSC                                                                                    14\n",
       "SAR                                                                                    13\n",
       "Other/Pleasure                                                                         11\n",
       "BigTow/Towing/Tug                                                                      11\n",
       "                                                                                     ... \n",
       "SAR/Unknown                                                                             1\n",
       "BigTow/Fishing/Tug                                                                      1\n",
       "Tanker/WIG                                                                              1\n",
       "Fishing/Pilot/Unknown                                                                   1\n",
       "BigTow/Fishing/Towing                                                                   1\n",
       "Dredging/Other/Unknown                                                                  1\n",
       "Cargo/Reserved/Unknown                                                                  1\n",
       "Tanker/Tug/WIG                                                                          1\n",
       "BigTow/Cargo/Dredging/Reserved                                                          1\n",
       "Fishing/Passenger                                                                       1\n",
       "BigTow/Resol-18/Towing                                                                  1\n",
       "Tanker/Towing                                                                           1\n",
       "MilOps/Other/Tanker                                                                     1\n",
       "HSC/Pleasure                                                                            1\n",
       "BigTow/Tug/WIG                                                                          1\n",
       "Cargo/Sailing                                                                           1\n",
       "Other/WIG                                                                               1\n",
       "Sailing/Spare                                                                           1\n",
       "BigTow/Reserved/Towing/Tug/WIG                                                          1\n",
       "Dredging/Other                                                                          1\n",
       "BigTow/Towing/Tug/Unknown                                                               1\n",
       "Medical                                                                                 1\n",
       "Fishing/Other/Towing                                                                    1\n",
       "AntiPol/BigTow/Fishing/Other/Passenger/Pilot/Pleasure/Resol-18/Sailing/Towing/Tug       1\n",
       "Reserved/Tanker/Towing/Tug                                                              1\n",
       "MilOps/Other/Towing                                                                     1\n",
       "Cargo/Tanker/Unknown                                                                    1\n",
       "Tender                                                                                  1\n",
       "MilOps/Reserved/Towing                                                                  1\n",
       "Other/Reserved/Tug                                                                      1\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vessels.type.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The challenge, however, is that several ships have travelled multiple segments, so there is not a one-to-one relationship between the rows of the two tables. The table of vessel information has a *one-to-many* relationship with the segments.\n",
    "\n",
    "In Pandas, we can combine tables according to the value of one or more *keys* that are used to identify rows, much like an index. Using a trivial example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   age  id\n",
       "0   22   0\n",
       "1   28   1\n",
       "2   27   2\n",
       "3   29   3"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame(dict(id=range(4), age=np.random.randint(18, 31, size=4)))\n",
    "df2 = pd.DataFrame(dict(id=list(range(3))+list(range(3)), \n",
    "                        score=np.random.random(size=6)))\n",
    "\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.552907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.308380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.341992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0.217117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.191608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>0.002176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id     score\n",
       "0   0  0.552907\n",
       "1   1  0.308380\n",
       "2   2  0.341992\n",
       "3   0  0.217117\n",
       "4   1  0.191608\n",
       "5   2  0.002176"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0.552907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0.217117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>0.308380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>0.191608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>0.341992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>0.002176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   age  id     score\n",
       "0   22   0  0.552907\n",
       "1   22   0  0.217117\n",
       "2   28   1  0.308380\n",
       "3   28   1  0.191608\n",
       "4   27   2  0.341992\n",
       "5   27   2  0.002176"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1, df2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that without any information about which column to use as a key, Pandas did the right thing and used the `id` column in both tables. Unless specified otherwise, `merge` will used any common column names as keys for merging the tables. \n",
    "\n",
    "Notice also that `id=3` from `df1` was omitted from the merged table. This is because, by default, `merge` performs an **inner join** on the tables, meaning that the merged table represents an intersection of the two tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0.552907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0.217117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>0.308380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>0.191608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>0.341992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>0.002176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   age  id     score\n",
       "0   22   0  0.552907\n",
       "1   22   0  0.217117\n",
       "2   28   1  0.308380\n",
       "3   28   1  0.191608\n",
       "4   27   2  0.341992\n",
       "5   27   2  0.002176\n",
       "6   29   3       NaN"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df1, df2, how='outer')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **outer join** above yields the union of the two tables, so all rows are represented, with missing values inserted as appropriate. One can also perform **right** and **left** joins to include all rows of the right or left table (*i.e.* first or second argument to `merge`), but not necessarily the other."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the two datasets that we wish to merge:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mmsi</th>\n",
       "      <th>name</th>\n",
       "      <th>transit</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_length</th>\n",
       "      <th>avg_sog</th>\n",
       "      <th>min_sog</th>\n",
       "      <th>max_sog</th>\n",
       "      <th>pdgt10</th>\n",
       "      <th>st_time</th>\n",
       "      <th>end_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Us Govt Ves</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>13.2</td>\n",
       "      <td>9.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>96.5</td>\n",
       "      <td>2009-02-10 16:03:00</td>\n",
       "      <td>2009-02-10 16:27:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mmsi         name  transit  segment  seg_length  avg_sog  min_sog  max_sog  \\\n",
       "0     1  Us Govt Ves        1        1         5.1     13.2      9.2     14.5   \n",
       "\n",
       "   pdgt10             st_time            end_time  \n",
       "0    96.5 2009-02-10 16:03:00 2009-02-10 16:27:00  "
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_names</th>\n",
       "      <th>names</th>\n",
       "      <th>sov</th>\n",
       "      <th>flag</th>\n",
       "      <th>flag_type</th>\n",
       "      <th>num_loas</th>\n",
       "      <th>loa</th>\n",
       "      <th>max_loa</th>\n",
       "      <th>num_types</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mmsi</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      num_names                                              names sov  \\\n",
       "mmsi                                                                     \n",
       "1             8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y   \n",
       "\n",
       "         flag flag_type  num_loas                                    loa  \\\n",
       "mmsi                                                                       \n",
       "1     Unknown   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0   \n",
       "\n",
       "      max_loa  num_types                             type  \n",
       "mmsi                                                       \n",
       "1       156.0          4  Dredging/MilOps/Reserved/Towing  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vessels.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we see that there is a `mmsi` value (a vessel identifier) in each table, but it is used as an index for the `vessels` table. In this case, we have to specify to join on the index for this table, and on the `mmsi` column for the other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "segments_merged = pd.merge(vessels, segments, left_index=True, right_on='mmsi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_names</th>\n",
       "      <th>names</th>\n",
       "      <th>sov</th>\n",
       "      <th>flag</th>\n",
       "      <th>flag_type</th>\n",
       "      <th>num_loas</th>\n",
       "      <th>loa</th>\n",
       "      <th>max_loa</th>\n",
       "      <th>num_types</th>\n",
       "      <th>type</th>\n",
       "      <th>...</th>\n",
       "      <th>name</th>\n",
       "      <th>transit</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_length</th>\n",
       "      <th>avg_sog</th>\n",
       "      <th>min_sog</th>\n",
       "      <th>max_sog</th>\n",
       "      <th>pdgt10</th>\n",
       "      <th>st_time</th>\n",
       "      <th>end_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Us Govt Ves</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>13.2</td>\n",
       "      <td>9.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>96.5</td>\n",
       "      <td>2009-02-10 16:03:00</td>\n",
       "      <td>2009-02-10 16:27:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Dredge Capt Frank</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>18.6</td>\n",
       "      <td>10.4</td>\n",
       "      <td>20.6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-06 14:31:00</td>\n",
       "      <td>2009-04-06 15:20:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4.3</td>\n",
       "      <td>16.2</td>\n",
       "      <td>10.3</td>\n",
       "      <td>20.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-06 14:36:00</td>\n",
       "      <td>2009-04-06 14:55:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.5</td>\n",
       "      <td>16.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-10 17:58:00</td>\n",
       "      <td>2009-04-10 18:34:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Dredge Capt Frank</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.6</td>\n",
       "      <td>16.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-10 17:59:00</td>\n",
       "      <td>2009-04-10 18:35:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_names                                              names sov     flag  \\\n",
       "0          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "1          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "2          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "3          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "4          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "\n",
       "  flag_type  num_loas                                    loa  max_loa  \\\n",
       "0   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "1   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "2   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "3   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "4   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "\n",
       "   num_types                             type         ...          \\\n",
       "0          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "1          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "2          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "3          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "4          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "\n",
       "                name transit  segment  seg_length  avg_sog  min_sog  max_sog  \\\n",
       "0        Us Govt Ves       1        1         5.1     13.2      9.2     14.5   \n",
       "1  Dredge Capt Frank       1        1        13.5     18.6     10.4     20.6   \n",
       "2      Us Gov Vessel       1        1         4.3     16.2     10.3     20.5   \n",
       "3      Us Gov Vessel       2        1         9.2     15.4     14.5     16.1   \n",
       "4  Dredge Capt Frank       2        1         9.2     15.4     14.6     16.2   \n",
       "\n",
       "   pdgt10             st_time            end_time  \n",
       "0    96.5 2009-02-10 16:03:00 2009-02-10 16:27:00  \n",
       "1   100.0 2009-04-06 14:31:00 2009-04-06 15:20:00  \n",
       "2   100.0 2009-04-06 14:36:00 2009-04-06 14:55:00  \n",
       "3   100.0 2009-04-10 17:58:00 2009-04-10 18:34:00  \n",
       "4   100.0 2009-04-10 17:59:00 2009-04-10 18:35:00  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case, the default inner join is suitable; we are not interested in observations from either table that do not have corresponding entries in the other. \n",
    "\n",
    "Notice that `mmsi` field that was an index on the `vessels` table is no longer an index on the merged table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we used the `merge` function to perform the merge; we could also have used the `merge` *method* for either of the tables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_names</th>\n",
       "      <th>names</th>\n",
       "      <th>sov</th>\n",
       "      <th>flag</th>\n",
       "      <th>flag_type</th>\n",
       "      <th>num_loas</th>\n",
       "      <th>loa</th>\n",
       "      <th>max_loa</th>\n",
       "      <th>num_types</th>\n",
       "      <th>type</th>\n",
       "      <th>...</th>\n",
       "      <th>name</th>\n",
       "      <th>transit</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_length</th>\n",
       "      <th>avg_sog</th>\n",
       "      <th>min_sog</th>\n",
       "      <th>max_sog</th>\n",
       "      <th>pdgt10</th>\n",
       "      <th>st_time</th>\n",
       "      <th>end_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Us Govt Ves</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>13.2</td>\n",
       "      <td>9.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>96.5</td>\n",
       "      <td>2009-02-10 16:03:00</td>\n",
       "      <td>2009-02-10 16:27:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Dredge Capt Frank</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>18.6</td>\n",
       "      <td>10.4</td>\n",
       "      <td>20.6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-06 14:31:00</td>\n",
       "      <td>2009-04-06 15:20:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4.3</td>\n",
       "      <td>16.2</td>\n",
       "      <td>10.3</td>\n",
       "      <td>20.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-06 14:36:00</td>\n",
       "      <td>2009-04-06 14:55:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Us Gov Vessel</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.5</td>\n",
       "      <td>16.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-10 17:58:00</td>\n",
       "      <td>2009-04-10 18:34:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>Dredge Capt Frank</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.6</td>\n",
       "      <td>16.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-10 17:59:00</td>\n",
       "      <td>2009-04-10 18:35:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_names                                              names sov     flag  \\\n",
       "0          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "1          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "2          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "3          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "4          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "\n",
       "  flag_type  num_loas                                    loa  max_loa  \\\n",
       "0   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "1   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "2   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "3   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "4   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "\n",
       "   num_types                             type         ...          \\\n",
       "0          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "1          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "2          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "3          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "4          4  Dredging/MilOps/Reserved/Towing         ...           \n",
       "\n",
       "                name transit  segment  seg_length  avg_sog  min_sog  max_sog  \\\n",
       "0        Us Govt Ves       1        1         5.1     13.2      9.2     14.5   \n",
       "1  Dredge Capt Frank       1        1        13.5     18.6     10.4     20.6   \n",
       "2      Us Gov Vessel       1        1         4.3     16.2     10.3     20.5   \n",
       "3      Us Gov Vessel       2        1         9.2     15.4     14.5     16.1   \n",
       "4  Dredge Capt Frank       2        1         9.2     15.4     14.6     16.2   \n",
       "\n",
       "   pdgt10             st_time            end_time  \n",
       "0    96.5 2009-02-10 16:03:00 2009-02-10 16:27:00  \n",
       "1   100.0 2009-04-06 14:31:00 2009-04-06 15:20:00  \n",
       "2   100.0 2009-04-06 14:36:00 2009-04-06 14:55:00  \n",
       "3   100.0 2009-04-10 17:58:00 2009-04-10 18:34:00  \n",
       "4   100.0 2009-04-10 17:59:00 2009-04-10 18:35:00  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vessels.merge(segments, left_index=True, right_on='mmsi').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Occasionally, there will be fields with the same in both tables that we do not wish to use to join the tables; they may contain different information, despite having the same name. In this case, Pandas will by default append suffixes `_x` and `_y` to the columns to uniquely identify them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_names</th>\n",
       "      <th>names</th>\n",
       "      <th>sov</th>\n",
       "      <th>flag</th>\n",
       "      <th>flag_type</th>\n",
       "      <th>num_loas</th>\n",
       "      <th>loa</th>\n",
       "      <th>max_loa</th>\n",
       "      <th>num_types</th>\n",
       "      <th>type_x</th>\n",
       "      <th>...</th>\n",
       "      <th>transit</th>\n",
       "      <th>segment</th>\n",
       "      <th>seg_length</th>\n",
       "      <th>avg_sog</th>\n",
       "      <th>min_sog</th>\n",
       "      <th>max_sog</th>\n",
       "      <th>pdgt10</th>\n",
       "      <th>st_time</th>\n",
       "      <th>end_time</th>\n",
       "      <th>type_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>13.2</td>\n",
       "      <td>9.2</td>\n",
       "      <td>14.5</td>\n",
       "      <td>96.5</td>\n",
       "      <td>2009-02-10 16:03:00</td>\n",
       "      <td>2009-02-10 16:27:00</td>\n",
       "      <td>foo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>18.6</td>\n",
       "      <td>10.4</td>\n",
       "      <td>20.6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-06 14:31:00</td>\n",
       "      <td>2009-04-06 15:20:00</td>\n",
       "      <td>foo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4.3</td>\n",
       "      <td>16.2</td>\n",
       "      <td>10.3</td>\n",
       "      <td>20.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-06 14:36:00</td>\n",
       "      <td>2009-04-06 14:55:00</td>\n",
       "      <td>foo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.5</td>\n",
       "      <td>16.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-10 17:58:00</td>\n",
       "      <td>2009-04-10 18:34:00</td>\n",
       "      <td>foo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7</td>\n",
       "      <td>42.0/48.0/57.0/90.0/138.0/154.0/156.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>4</td>\n",
       "      <td>Dredging/MilOps/Reserved/Towing</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9.2</td>\n",
       "      <td>15.4</td>\n",
       "      <td>14.6</td>\n",
       "      <td>16.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2009-04-10 17:59:00</td>\n",
       "      <td>2009-04-10 18:35:00</td>\n",
       "      <td>foo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_names                                              names sov     flag  \\\n",
       "0          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "1          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "2          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "3          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "4          8  Bil Holman Dredge/Dredge Capt Frank/Emo/Offsho...   Y  Unknown   \n",
       "\n",
       "  flag_type  num_loas                                    loa  max_loa  \\\n",
       "0   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "1   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "2   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "3   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "4   Unknown         7  42.0/48.0/57.0/90.0/138.0/154.0/156.0    156.0   \n",
       "\n",
       "   num_types                           type_x  ...    transit segment  \\\n",
       "0          4  Dredging/MilOps/Reserved/Towing  ...          1       1   \n",
       "1          4  Dredging/MilOps/Reserved/Towing  ...          1       1   \n",
       "2          4  Dredging/MilOps/Reserved/Towing  ...          1       1   \n",
       "3          4  Dredging/MilOps/Reserved/Towing  ...          2       1   \n",
       "4          4  Dredging/MilOps/Reserved/Towing  ...          2       1   \n",
       "\n",
       "   seg_length  avg_sog  min_sog  max_sog  pdgt10             st_time  \\\n",
       "0         5.1     13.2      9.2     14.5    96.5 2009-02-10 16:03:00   \n",
       "1        13.5     18.6     10.4     20.6   100.0 2009-04-06 14:31:00   \n",
       "2         4.3     16.2     10.3     20.5   100.0 2009-04-06 14:36:00   \n",
       "3         9.2     15.4     14.5     16.1   100.0 2009-04-10 17:58:00   \n",
       "4         9.2     15.4     14.6     16.2   100.0 2009-04-10 17:59:00   \n",
       "\n",
       "             end_time type_y  \n",
       "0 2009-02-10 16:27:00    foo  \n",
       "1 2009-04-06 15:20:00    foo  \n",
       "2 2009-04-06 14:55:00    foo  \n",
       "3 2009-04-10 18:34:00    foo  \n",
       "4 2009-04-10 18:35:00    foo  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segments['type'] = 'foo'\n",
    "pd.merge(vessels, segments, left_index=True, right_on='mmsi').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This behavior can be overridden by specifying a `suffixes` argument, containing a list of the suffixes to be used for the columns of the left and right columns, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Concatenation\n",
    "\n",
    "A common data manipulation is appending rows or columns to a dataset that already conform to the dimensions of the exsiting rows or colums, respectively. In NumPy, this is done either with `concatenate` or the convenience \"functions\" `c_` and `r_`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.40009841,  0.91622185,  0.78812062,  0.00544846,  0.89776734,\n",
       "        0.6663524 ,  0.02489468,  0.22105227,  0.40472224,  0.61640489])"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.concatenate([np.random.random(5), np.random.random(5)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.99156254,  0.01896024,  0.36672654,  0.25266127,  0.37461893,\n",
       "        0.8997894 ,  0.04967815,  0.50456231,  0.41999845,  0.82115445])"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.r_[np.random.random(5), np.random.random(5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.96444945,  0.7203291 ],\n",
       "       [ 0.72480403,  0.68867113],\n",
       "       [ 0.63456914,  0.58312985],\n",
       "       [ 0.41537736,  0.01793207],\n",
       "       [ 0.07328689,  0.63548117]])"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.c_[np.random.random(5), np.random.random(5)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Notice that `c_` and `r_` are not really functions at all, since it is performing some sort of indexing operation, rather than being called. They are actually *class instances*, but they are here behaving mostly like functions. Don't think about this too hard; just know that they are there."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This operation is also called *binding* or *stacking*.\n",
    "\n",
    "With Pandas' indexed data structures, there are additional considerations as the overlap in index values between two data structures affects how they are concatenate.\n",
    "\n",
    "Lets import two microbiome datasets, each consisting of counts of microorganiams from a particular patient. We will use the first column of each dataset as the index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((272, 1), (288, 1))"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mb1 = pd.read_excel('Data/microbiome/MID1.xls', 'Sheet 1', index_col=0, header=None)\n",
    "mb2 = pd.read_excel('Data/microbiome/MID2.xls', 'Sheet 1', index_col=0, header=None)\n",
    "mb1.shape, mb2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    1\n",
       "0                                                    \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  7\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  2\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  3\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  3\n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  7"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mb1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's give the index and columns meaningful labels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mb1.columns = mb2.columns = ['Count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mb1.index.name = mb2.index.name = 'Taxon'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Count\n",
       "Taxon                                                    \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3\n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mb1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The index of these data is the unique biological classification of each organism, beginning with *domain*, *phylum*, *class*, and for some organisms, going all the way down to the genus level.\n",
    "\n",
    "![classification](http://upload.wikimedia.org/wikipedia/commons/thumb/a/a5/Biological_classification_L_Pengo_vflip.svg/150px-Biological_classification_L_Pengo_vflip.svg.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera',\n",
       "       'Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus',\n",
       "       'Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus'],\n",
       "      dtype='object', name='Taxon')"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mb1.index[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mb1.index.is_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we concatenate along `axis=0` (the default), we will obtain another data frame with the the rows concatenated:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(560, 1)"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], axis=0).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, the index is no longer unique, due to overlap between the two DataFrames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], axis=0).index.is_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concatenating along `axis=1` will concatenate column-wise, but respecting the indices of the two DataFrames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(438, 2)"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], axis=1).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Acidilobaceae Acidilobus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Caldisphaeraceae Caldisphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Sulfophobococcus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermosphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Count  Count\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    NaN    2.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    NaN   14.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    7.0   23.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    1.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    2.0"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], axis=1).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we are only interested in taxa that are included in both DataFrames, we can specify a `join=inner` argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Count  Count\n",
       "Taxon                                                           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7     23\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2      2\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3     10\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3      9\n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7      9"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], axis=1, join='inner').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we wanted to use the second table to *fill values* absent from the first table, we could use `combine_first`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Acidilobaceae Acidilobus</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Caldisphaeraceae Caldisphaera</th>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Sulfophobococcus</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermosphaera</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Count\n",
       "Taxon                                                    \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    2.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...   14.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    7.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    1.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    2.0"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mb1.combine_first(mb2).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also create a hierarchical index based on keys identifying the original tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">patient1</th>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             Count\n",
       "         Taxon                                                    \n",
       "patient1 Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7\n",
       "         Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2\n",
       "         Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3\n",
       "         Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3\n",
       "         Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], keys=['patient1', 'patient2']).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([mb1, mb2], keys=['patient1', 'patient2']).index.is_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, you can pass keys to the concatenation by supplying the DataFrames (or Series) as a dict, resulting in a \"wide\" format table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>patient1</th>\n",
       "      <th>patient2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Acidilobaceae Acidilobus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Caldisphaeraceae Caldisphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Sulfophobococcus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermosphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   patient1 patient2\n",
       "                                                      Count    Count\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...      NaN      2.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...      NaN     14.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7.0     23.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      NaN      1.0\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      NaN      2.0"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat(dict(patient1=mb1, patient2=mb2), axis=1).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want `concat` to work like `numpy.concatanate`, you may provide the `ignore_index=True` argument."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise\n",
    "\n",
    "In the *data/microbiome* subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10th file that describes the content of each. Write code that imports each of the data spreadsheets and combines them into a single `DataFrame`, adding the identifying information from the metadata spreadsheet as columns in the combined `DataFrame`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COUNT</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>...</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Acidilobaceae Acidilobus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Acidilobales Caldisphaeraceae Caldisphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>15.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>14.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Stetteria</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Sulfophobococcus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermodiscus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Thermosphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Hyperthermus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrodictium</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "      <td>6.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Sulfurisphaera</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermoproteaceae Caldivirga</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermoproteaceae Thermocladium</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>7.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanomicrobiaceae Methanomicrobium</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanomicrobiaceae Methanosphaerula</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanomicrobiales_incertae_sedis Methanolinea</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanomicrobiales Methanospirillaceae Methanospirillum</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanimicrococcus</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanosalsum</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methermicoccaceae Methermicoccus</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Ferroglobus</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Geoglobus</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haladaptatus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halogeometricum</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halopiger</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloplanus</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloquadratum</th>\n",
       "      <td>4.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Rudaea</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Silanimonas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Stenotrophomonas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Wohlfahrtiimonas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Xanthomonas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Xylella</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Spirochaetes\" Spirochaetes Spirochaetales Spirochaetales_incertae_sedis Exilispira</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminiphilus</th>\n",
       "      <td>2.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminomonas</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Thermovirga</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Anaeroplasmatales Anaeroplasmataceae Asteroleplasma</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Haloplasmatales Haloplasmataceae Haloplasma</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Caldimicrobium</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Thermodesulfatator</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Geotoga</th>\n",
       "      <td>7.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>12.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "      <td>43.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Kosmotoga</th>\n",
       "      <td>9.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>7.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "      <td>16.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Alterococcus</th>\n",
       "      <td>1.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Opitutus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Puniceicoccales Puniceicoccaceae Cerasicoccus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Puniceicoccales Puniceicoccaceae Coraliomargarita</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Puniceicoccales Puniceicoccaceae Puniceicoccus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Subdivision3   Subdivision3_genera_incertae_sedis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Verrucomicrobiae Verrucomicrobiales Verrucomicrobiaceae Persicirhabdus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Verrucomicrobiae Verrucomicrobiales Verrucomicrobiaceae Verrucomicrobium</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Bangiophyceae</th>\n",
       "      <td>2.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Chlorarachniophyceae</th>\n",
       "      <td>85.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>Control 1</td>\n",
       "      <td>stool</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MID8</td>\n",
       "      <td>NEC 2</td>\n",
       "      <td>stool</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Streptophyta</th>\n",
       "      <td>1388.0</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>54.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria OD1    OD1_genera_incertae_sedis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria OP11    OP11_genera_incertae_sedis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>MID3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria TM7    TM7_genera_incertae_sedis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MID2</td>\n",
       "      <td>NEC 1</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>775 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     COUNT BARCODE  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     7.0    MID1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     2.0    MID1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...     3.0    MID1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...     3.0    MID1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...     NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     7.0    MID1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     1.0    MID1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     1.0    MID1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...     1.0    MID1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...     1.0    MID1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...     NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...     1.0    MID1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...     4.0    MID1   \n",
       "...                                                    ...     ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN     NaN   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...     1.0    MID1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...     2.0    MID1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...     1.0    MID1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...     1.0    MID1   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...     NaN     NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...     1.0    MID1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...     1.0    MID1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...     NaN     NaN   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...     7.0    MID1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...     9.0    MID1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...     1.0    MID1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...     NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...     NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...     NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...     NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...     NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...     NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...     NaN     NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...     2.0    MID1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    85.0    MID1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  1388.0    MID1   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis              NaN     NaN   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis            NaN     NaN   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis              NaN     NaN   \n",
       "\n",
       "                                                                 GROUP SAMPLE  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...                 NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...                 NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL    NaN   \n",
       "...                                                                ...    ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...                 NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...                 NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...                 NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...                 NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...                 NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...                 NaN    NaN   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...                 NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...                 NaN    NaN   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...                 NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...                 NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...                 NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...                 NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...                 NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...                 NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...                 NaN    NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  EXTRACTION CONTROL    NaN   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis                          NaN    NaN   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis                        NaN    NaN   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis                          NaN    NaN   \n",
       "\n",
       "                                                    COUNT BARCODE  GROUP  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    2.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...   14.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...   23.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    1.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    2.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    1.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    2.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...   10.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...   11.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    9.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    5.0    MID2  NEC 1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    3.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    9.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    1.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...   12.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    2.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...    4.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    3.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    1.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    4.0    MID2  NEC 1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    3.0    MID2  NEC 1   \n",
       "...                                                   ...     ...    ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...    NaN     NaN    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    2.0    MID2  NEC 1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    2.0    MID2  NEC 1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN     NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...    NaN     NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...   15.0    MID2  NEC 1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    1.0    MID2  NEC 1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    3.0    MID2  NEC 1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...   15.0    MID2  NEC 1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...   22.0    MID2  NEC 1   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...    NaN     NaN    NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    NaN     NaN    NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    1.0    MID2  NEC 1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    2.0    MID2  NEC 1   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis             NaN     NaN    NaN   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis           NaN     NaN    NaN   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis             2.0    MID2  NEC 1   \n",
       "\n",
       "                                                    SAMPLE  COUNT BARCODE  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  tissue    1.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  tissue   15.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  tissue   14.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  tissue    4.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN    1.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  tissue    1.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  tissue    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...     NaN    3.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  tissue    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  tissue    4.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  tissue    2.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  tissue    5.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  tissue    3.0    MID3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  tissue    1.0    MID3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  tissue    7.0    MID3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  tissue    4.0    MID3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN    1.0    MID3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  tissue    2.0    MID3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...     NaN    1.0    MID3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  tissue    1.0    MID3   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  tissue    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...     NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  tissue    3.0    MID3   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  tissue    1.0    MID3   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...     NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  tissue    3.0    MID3   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  tissue    2.0    MID3   \n",
       "...                                                    ...    ...     ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN    1.0    MID3   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN    1.0    MID3   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...     NaN    NaN     NaN   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...     NaN    NaN     NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  tissue    2.0    MID3   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  tissue    3.0    MID3   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...     NaN    NaN     NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...     NaN    NaN     NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  tissue    2.0    MID3   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  tissue    NaN     NaN   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  tissue    1.0    MID3   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  tissue   12.0    MID3   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  tissue    7.0    MID3   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...     NaN    1.0    MID3   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...     NaN    1.0    MID3   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...     NaN    1.0    MID3   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...     NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...     NaN    2.0    MID3   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...     NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...     NaN    2.0    MID3   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...     NaN    1.0    MID3   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...     NaN    1.0    MID3   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  tissue    4.0    MID3   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  tissue   54.0    MID3   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis              NaN    1.0    MID3   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis            NaN   26.0    MID3   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis           tissue    NaN     NaN   \n",
       "\n",
       "                                                    ...        GROUP SAMPLE  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  ...          NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...          NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...          NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...          NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  ...    Control 1  stool   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  ...          NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  ...          NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  ...    Control 1  stool   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  ...          NaN    NaN   \n",
       "...                                                 ...          ...    ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  ...    Control 1  stool   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  ...          NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  ...    Control 1  stool   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  ...          NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  ...    Control 1  stool   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...  ...    Control 1  stool   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...  ...          NaN    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  ...    Control 1  stool   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  ...    Control 1  stool   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  ...          NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  ...          NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  ...    Control 1  stool   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  ...    Control 1  stool   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  ...    Control 1  stool   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  ...    Control 1  stool   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...  ...          NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  ...    Control 1  stool   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...  ...    Control 1  stool   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  ...    Control 1  stool   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  ...    Control 1  stool   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  ...          NaN    NaN   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis           ...          NaN    NaN   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis         ...          NaN    NaN   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis           ...          NaN    NaN   \n",
       "\n",
       "                                                    COUNT BARCODE  GROUP  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...    1.0    MID8  NEC 2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    NaN     NaN    NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN     NaN    NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN     NaN    NaN   \n",
       "...                                                   ...     ...    ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN     NaN    NaN   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...    NaN     NaN    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN     NaN    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN     NaN    NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN     NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...    NaN     NaN    NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...    NaN     NaN    NaN   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    NaN     NaN    NaN   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    1.0    MID8  NEC 2   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...    4.0    MID8  NEC 2   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...    2.0    MID8  NEC 2   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...    NaN     NaN    NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...    NaN     NaN    NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    1.0    MID8  NEC 2   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    4.0    MID8  NEC 2   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    NaN     NaN    NaN   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis             NaN     NaN    NaN   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis           NaN     NaN    NaN   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis             NaN     NaN    NaN   \n",
       "\n",
       "                                                   SAMPLE  COUNT BARCODE  \\\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...    NaN    1.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN   16.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    2.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    5.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...    NaN    NaN     NaN   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  stool    6.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...    NaN    2.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    NaN    5.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    NaN    3.0    MID9   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...    NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN   14.0    MID9   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    1.0    MID9   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    2.0    MID9   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    1.0    MID9   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    6.0    MID9   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...    NaN    3.0    MID9   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...    NaN    3.0    MID9   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...    NaN    3.0    MID9   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN    NaN     NaN   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN    4.0    MID9   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...    NaN    1.0    MID9   \n",
       "...                                                   ...    ...     ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN    NaN     NaN   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...    NaN    NaN     NaN   \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...    NaN    NaN     NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN    NaN     NaN   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN    3.0    MID9   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...    NaN    NaN     NaN   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...    NaN    1.0    MID9   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...    NaN   11.0    MID9   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    NaN    1.0    MID9   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  stool    1.0    MID9   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  stool   43.0    MID9   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  stool   16.0    MID9   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    NaN    1.0    MID9   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...    NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...    NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...    NaN    NaN     NaN   \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...    NaN    NaN     NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  stool    NaN     NaN   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  stool    1.0    MID9   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...    NaN   10.0    MID9   \n",
       "Bacteria OD1    OD1_genera_incertae_sedis             NaN    NaN     NaN   \n",
       "Bacteria OP11    OP11_genera_incertae_sedis           NaN    NaN     NaN   \n",
       "Bacteria TM7    TM7_genera_incertae_sedis             NaN    NaN     NaN   \n",
       "\n",
       "                                                        GROUP SAMPLE  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...        NaN    NaN  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Acidilobal...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        NaN    NaN  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        NaN    NaN  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        NaN    NaN  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        NaN    NaN  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        NaN    NaN  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  Control 2  stool  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...        NaN    NaN  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        NaN    NaN  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        NaN    NaN  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        NaN    NaN  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        NaN    NaN  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...        NaN    NaN  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  Control 2  stool  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter...  Control 2  stool  \n",
       "...                                                       ...    ...  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        NaN    NaN  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        NaN    NaN  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        NaN    NaN  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        NaN    NaN  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        NaN    NaN  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X...        NaN    NaN  \n",
       "Bacteria \"Spirochaetes\" Spirochaetes Spirochaet...        NaN    NaN  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        NaN    NaN  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...  Control 2  stool  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista...        NaN    NaN  \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...  Control 2  stool  \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...  Control 2  stool  \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  Control 2  stool  \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob...  Control 2  stool  \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  Control 2  stool  \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale...  Control 2  stool  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...        NaN    NaN  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...  Control 2  stool  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...        NaN    NaN  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...        NaN    NaN  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Puniceicocc...        NaN    NaN  \n",
       "Bacteria \"Verrucomicrobia\" Subdivision3   Subdi...        NaN    NaN  \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...        NaN    NaN  \n",
       "Bacteria \"Verrucomicrobia\" Verrucomicrobiae Ver...        NaN    NaN  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...        NaN    NaN  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  Control 2  stool  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla...  Control 2  stool  \n",
       "Bacteria OD1    OD1_genera_incertae_sedis                 NaN    NaN  \n",
       "Bacteria OP11    OP11_genera_incertae_sedis               NaN    NaN  \n",
       "Bacteria TM7    TM7_genera_incertae_sedis                 NaN    NaN  \n",
       "\n",
       "[775 rows x 36 columns]"
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write solution here\n",
    "import glob\n",
    "import os.path as osp\n",
    "\n",
    "#Read spreadsheets\n",
    "targetDir = \"Data/microbiome\"\n",
    "allFiles = glob.glob(targetDir + \"/MID*.xls\")\n",
    "list_ = []\n",
    "for file_ in allFiles:\n",
    "    basename = osp.splitext(osp.basename(file_))[0]\n",
    "    #Read file\n",
    "    df = pd.read_excel(file_, index_col=0, header=None, names=['COUNT'])\n",
    "    df[df_metadata.index.name] = basename\n",
    "    for (column, value) in zip(df_metadata.columns, df_metadata.loc[basename]):\n",
    "        df[column] = value\n",
    "    #Swap columns (to have the count after each information)\n",
    "    df_cols = df.columns.tolist()\n",
    "    df = df[df_cols[1:] + df_cols[:1]]\n",
    "    #Append to the list\n",
    "    list_.append(df)\n",
    "\n",
    "df_result = pd.concat(list_, axis=1)\n",
    "df_result\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reshaping DataFrame objects\n",
    "\n",
    "In the context of a single DataFrame, we are often interested in re-arranging the layout of our data. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This dataset is from Table 6.9 of [Statistical Methods for the Analysis of Repeated Measurements](http://www.amazon.com/Statistical-Methods-Analysis-Repeated-Measurements/dp/0387953701) by Charles S. Davis, pp. 161-163 (Springer, 2002). These data are from a multicenter, randomized controlled trial of botulinum toxin type B (BotB) in patients with cervical dystonia from nine U.S. sites.\n",
    "\n",
    "* Randomized to placebo (N=36), 5000 units of BotB (N=36), 10,000 units of BotB (N=37)\n",
    "* Response variable: total score on Toronto Western Spasmodic Torticollis Rating Scale (TWSTRS), measuring severity, pain, and disability of cervical dystonia (high scores mean more impairment)\n",
    "* TWSTRS measured at baseline (week 0) and weeks 2, 4, 8, 12, 16 after treatment began"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>obs</th>\n",
       "      <th>week</th>\n",
       "      <th>site</th>\n",
       "      <th>id</th>\n",
       "      <th>treat</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>twstrs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5000U</td>\n",
       "      <td>65</td>\n",
       "      <td>F</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5000U</td>\n",
       "      <td>65</td>\n",
       "      <td>F</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5000U</td>\n",
       "      <td>65</td>\n",
       "      <td>F</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5000U</td>\n",
       "      <td>65</td>\n",
       "      <td>F</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5000U</td>\n",
       "      <td>65</td>\n",
       "      <td>F</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient  obs  week  site  id  treat  age sex  twstrs\n",
       "0        1    1     0     1   1  5000U   65   F      32\n",
       "1        1    2     2     1   1  5000U   65   F      30\n",
       "2        1    3     4     1   1  5000U   65   F      24\n",
       "3        1    4     8     1   1  5000U   65   F      37\n",
       "4        1    5    12     1   1  5000U   65   F      39"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia = pd.read_csv(\"Data/cdystonia.csv\", index_col=None)\n",
    "cdystonia.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This dataset includes repeated measurements of the same individuals (longitudinal data). Its possible to present such information in (at least) two ways: showing each repeated measurement in their own row, or in multiple columns representing multiple measurements.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `stack` method rotates the data frame so that columns are represented in rows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    patient        1\n",
       "     obs            1\n",
       "     week           0\n",
       "     site           1\n",
       "     id             1\n",
       "     treat      5000U\n",
       "     age           65\n",
       "     sex            F\n",
       "     twstrs        32\n",
       "1    patient        1\n",
       "     obs            2\n",
       "     week           2\n",
       "     site           1\n",
       "     id             1\n",
       "     treat      5000U\n",
       "     age           65\n",
       "     sex            F\n",
       "     twstrs        30\n",
       "2    patient        1\n",
       "     obs            3\n",
       "     week           4\n",
       "     site           1\n",
       "     id             1\n",
       "     treat      5000U\n",
       "     age           65\n",
       "     sex            F\n",
       "     twstrs        24\n",
       "3    patient        1\n",
       "     obs            4\n",
       "     week           8\n",
       "                ...  \n",
       "627  age           57\n",
       "     sex            M\n",
       "     twstrs        38\n",
       "628  patient      109\n",
       "     obs            4\n",
       "     week           8\n",
       "     site           9\n",
       "     id            11\n",
       "     treat      5000U\n",
       "     age           57\n",
       "     sex            M\n",
       "     twstrs        33\n",
       "629  patient      109\n",
       "     obs            5\n",
       "     week          12\n",
       "     site           9\n",
       "     id            11\n",
       "     treat      5000U\n",
       "     age           57\n",
       "     sex            M\n",
       "     twstrs        36\n",
       "630  patient      109\n",
       "     obs            6\n",
       "     week          16\n",
       "     site           9\n",
       "     id            11\n",
       "     treat      5000U\n",
       "     age           57\n",
       "     sex            M\n",
       "     twstrs        51\n",
       "dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stacked = cdystonia.stack()\n",
    "stacked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To complement this, `unstack` pivots from rows back to columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "stacked.unstack().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this dataset, it makes sense to create a hierarchical index based on the patient and observation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia2 = cdystonia.set_index(['patient','obs'])\n",
    "cdystonia2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia2.index.is_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we want to transform this data so that repeated measurements are in columns, we can `unstack` the `twstrs` measurements according to `obs`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "twstrs_wide = cdystonia2['twstrs'].unstack('obs')\n",
    "twstrs_wide.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_wide = (cdystonia[['patient','site','id','treat','age','sex']]\n",
    "                  .drop_duplicates()\n",
    "                  .merge(twstrs_wide, right_index=True, left_on='patient', how='inner')\n",
    "                  .head())\n",
    "cdystonia_wide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A slightly cleaner way of doing this is to set the patient-level information as an index before unstacking:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "(cdystonia.set_index(['patient','site','id','treat','age','sex','week'])['twstrs']\n",
    "     .unstack('week').head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To convert our \"wide\" format back to long, we can use the `melt` function, appropriately parameterized. This function is useful for `DataFrame`s where one\n",
    "or more columns are identifier variables (`id_vars`), with the remaining columns being measured variables (`value_vars`). The measured variables are \"unpivoted\" to\n",
    "the row axis, leaving just two non-identifier columns, a *variable* and its corresponding *value*, which can both be renamed using optional arguments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.melt(cdystonia_wide, id_vars=['patient','site','id','treat','age','sex'], \n",
    "        var_name='obs', value_name='twsters').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This illustrates the two formats for longitudinal data: **long** and **wide** formats. Its typically better to store data in long format because additional data can be included as additional rows in the database, while wide format requires that the entire database schema be altered by adding columns to every row as data are collected.\n",
    "\n",
    "The preferable format for analysis depends entirely on what is planned for the data, so it is imporant to be able to move easily between them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pivoting\n",
    "\n",
    "The `pivot` method allows a DataFrame to be transformed easily between long and wide formats in the same way as a pivot table is created in a spreadsheet. It takes three arguments: `index`, `columns` and `values`, corresponding to the DataFrame index (the row headers), columns and cell values, respectively.\n",
    "\n",
    "For example, we may want the `twstrs` variable (the response variable) in wide format according to patient, as we saw with the unstacking method above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia.pivot(index='patient', columns='obs', values='twstrs').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we omit the `values` argument, we get a `DataFrame` with hierarchical columns, just as when we applied `unstack` to the hierarchically-indexed table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia.pivot('patient', 'obs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A related method, `pivot_table`, creates a spreadsheet-like table with a hierarchical index, and allows the values of the table to be populated using an arbitrary aggregation function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia.pivot_table(index=['site', 'treat'], columns='week', values='twstrs', \n",
    "                      aggfunc=max).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a simple cross-tabulation of group frequencies, the `crosstab` function (not a method) aggregates counts of data according to factors in rows and columns. The factors may be hierarchical if desired."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.crosstab(cdystonia.sex, cdystonia.site)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data transformation\n",
    "\n",
    "There are a slew of additional operations for DataFrames that we would collectively refer to as \"transformations\" which include tasks such as removing duplicate values, replacing values, and grouping values."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dealing with duplicates\n",
    "\n",
    "We can easily identify and remove duplicate values from `DataFrame` objects. For example, say we want to removed ships from our `vessels` dataset that have the same name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vessels.duplicated(subset='names')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "vessels.drop_duplicates(['names'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Value replacement\n",
    "\n",
    "Frequently, we get data columns that are encoded as strings that we wish to represent numerically for the purposes of including it in a quantitative analysis. For example, consider the treatment variable in the cervical dystonia dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia.treat.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A logical way to specify these numerically is to change them to integer values, perhaps using \"Placebo\" as a baseline value. If we create a dict with the original values as keys and the replacements as values, we can pass it to the `map` method to implement the changes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "treatment_map = {'Placebo': 0, '5000U': 1, '10000U': 2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia['treatment'] = cdystonia.treat.map(treatment_map)\n",
    "cdystonia.treatment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternately, if we simply want to replace particular values in a `Series` or `DataFrame`, we can use the `replace` method. \n",
    "\n",
    "An example where replacement is useful is dealing with zeros in certain transformations. For example, if we try to take the log of a set of values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vals = pd.Series([float(i)**10 for i in range(10)])\n",
    "vals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "np.log(vals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In such situations, we can replace the zero with a value so small that it makes no difference to the ensuing analysis. We can do this with `replace`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vals = vals.replace(0, 1e-6)\n",
    "np.log(vals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also perform the same replacement that we used `map` for with `replace`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia2.treat.replace({'Placebo': 0, '5000U': 1, '10000U': 2})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inidcator variables\n",
    "\n",
    "For some statistical analyses (*e.g.* regression models or analyses of variance), categorical or group variables need to be converted into columns of indicators--zeros and ones--to create a so-called **design matrix**. The Pandas function `get_dummies` (indicator variables are also known as *dummy variables*) makes this transformation straightforward.\n",
    "\n",
    "Let's consider the DataFrame containing the ships corresponding to the transit segments on the eastern seaboard. The `type` variable denotes the class of vessel; we can create a matrix of indicators for this. For simplicity, lets filter out the 5 most common types of ships:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "top5 = vessels.type.isin(vessels.type.value_counts().index[:5])\n",
    "top5.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vessels5 = vessels[top5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.get_dummies(vessels5.type).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Categorical Data\n",
    "\n",
    "Pandas provides a convenient `dtype` for reprsenting categorical (factor) data, called `category`. \n",
    "\n",
    "For example, the `treat` column in the cervical dystonia dataset represents three treatment levels in a clinical trial, and is imported by default as an `object` type, since it is a mixture of string characters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    5000U\n",
       "1    5000U\n",
       "2    5000U\n",
       "3    5000U\n",
       "4    5000U\n",
       "Name: treat, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia.treat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can convert this to a `category` type either by the `Categorical` constructor, or casting the column using `astype`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[5000U, 5000U, 5000U, 5000U, 5000U, ..., 5000U, 5000U, 5000U, 5000U, 5000U]\n",
       "Length: 631\n",
       "Categories (3, object): [10000U, 5000U, Placebo]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.Categorical(cdystonia.treat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia['treat'] = cdystonia.treat.astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count        631\n",
       "unique         3\n",
       "top       10000U\n",
       "freq         213\n",
       "Name: treat, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia.treat.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default the Categorical type represents an unordered categorical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['10000U', '5000U', 'Placebo'], dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia.treat.cat.categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, an ordering can be imposed. The order is lexical by default, but will assume the order of the listed categories to be the desired order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia.treat.cat.categories = ['Placebo', '5000U', '10000U']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    5000U\n",
       "1    5000U\n",
       "2    5000U\n",
       "3    5000U\n",
       "4    5000U\n",
       "dtype: category\n",
       "Categories (3, object): [10000U < 5000U < Placebo]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia.treat.cat.as_ordered().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The important difference between the `category` type and the `object` type is that `category` is represented by an underlying array of integers, which is then mapped to character labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      1\n",
       "1      1\n",
       "2      1\n",
       "3      1\n",
       "4      1\n",
       "5      1\n",
       "6      0\n",
       "7      0\n",
       "8      0\n",
       "9      0\n",
       "10     0\n",
       "11     0\n",
       "12     1\n",
       "13     1\n",
       "14     1\n",
       "15     1\n",
       "16     1\n",
       "17     1\n",
       "18     2\n",
       "19     2\n",
       "20     2\n",
       "21     2\n",
       "22     0\n",
       "23     0\n",
       "24     0\n",
       "25     0\n",
       "26     0\n",
       "27     0\n",
       "28     0\n",
       "29     0\n",
       "      ..\n",
       "601    0\n",
       "602    0\n",
       "603    0\n",
       "604    2\n",
       "605    2\n",
       "606    2\n",
       "607    2\n",
       "608    2\n",
       "609    2\n",
       "610    1\n",
       "611    1\n",
       "612    1\n",
       "613    1\n",
       "614    1\n",
       "615    1\n",
       "616    0\n",
       "617    0\n",
       "618    0\n",
       "619    0\n",
       "620    0\n",
       "621    0\n",
       "622    0\n",
       "623    0\n",
       "624    0\n",
       "625    0\n",
       "626    1\n",
       "627    1\n",
       "628    1\n",
       "629    1\n",
       "630    1\n",
       "dtype: int8"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia.treat.cat.codes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that these are 8-bit integers, which are essentially single bytes of data, making memory usage lower.\n",
    "\n",
    "There is also a performance benefit. Consider an operation such as calculating the total segment lengths for each ship in the `segments` table (this is also a preview of pandas' `groupby` operation!):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'segments' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-727a1d90c5f1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmagic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'time segments.groupby(segments.name).seg_length.sum().sort_values(ascending=False, inplace=False).head()'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/home/hparmantier/anaconda3/lib/python3.5/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mmagic\u001b[0;34m(self, arg_s)\u001b[0m\n\u001b[1;32m   2156\u001b[0m         \u001b[0mmagic_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmagic_arg_s\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marg_s\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartition\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m' '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2157\u001b[0m         \u001b[0mmagic_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmagic_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprefilter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mESC_MAGIC\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2158\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_line_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmagic_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2159\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2160\u001b[0m     \u001b[0;31m#-------------------------------------------------------------------------\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/hparmantier/anaconda3/lib/python3.5/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_line_magic\u001b[0;34m(self, magic_name, line)\u001b[0m\n\u001b[1;32m   2077\u001b[0m                 \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'local_ns'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstack_depth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf_locals\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2078\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2079\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2080\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2081\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<decorator-gen-59>\u001b[0m in \u001b[0;36mtime\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n",
      "\u001b[0;32m/home/hparmantier/anaconda3/lib/python3.5/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    187\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 188\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    189\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    190\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/hparmantier/anaconda3/lib/python3.5/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtime\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m   1174\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'eval'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1175\u001b[0m             \u001b[0mst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclock2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1176\u001b[0;31m             \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0meval\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mglob\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlocal_ns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1177\u001b[0m             \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclock2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1178\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<timed eval>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'segments' is not defined"
     ]
    }
   ],
   "source": [
    "%time segments.groupby(segments.name).seg_length.sum().sort_values(ascending=False, inplace=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'segments' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-74a3b749262f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msegments\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msegments\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'category'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'segments' is not defined"
     ]
    }
   ],
   "source": [
    "segments['name'] = segments.name.astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%time segments.groupby(segments.name).seg_length.sum().sort_values(ascending=False, inplace=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hence, we get a considerable speedup simply by using the appropriate `dtype` for our data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discretization\n",
    "\n",
    "Pandas' `cut` function can be used to group continuous or countable data in to bins. Discretization is generally a very **bad idea** for statistical analysis, so use this function responsibly!\n",
    "\n",
    "Lets say we want to bin the ages of the cervical dystonia patients into a smaller number of groups:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    631.000000\n",
       "mean      55.616482\n",
       "std       12.123910\n",
       "min       26.000000\n",
       "25%       46.000000\n",
       "50%       56.000000\n",
       "75%       65.000000\n",
       "max       83.000000\n",
       "Name: age, dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdystonia.age.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's transform these data into decades, beginnnig with individuals in their 20's and ending with those in their 80's:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.cut(cdystonia.age, [20,30,40,50,60,70,80,90])[:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The parentheses indicate an open interval, meaning that the interval includes values up to but *not including* the endpoint, whereas the square bracket is a closed interval, where the endpoint is included in the interval. We can switch the closure to the left side by setting the `right` flag to `False`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.cut(cdystonia.age, [20,30,40,50,60,70,80,90], right=False)[:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the data are now **ordinal**, rather than numeric, we can give them labels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.cut(cdystonia.age, [20,40,60,80,90], labels=['young','middle-aged','old','really old'])[:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A related function `qcut` uses empirical quantiles to divide the data. If, for example, we want the quartiles -- (0-25%], (25-50%], (50-70%], (75-100%] -- we can just specify 4 intervals, which will be equally-spaced by default:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.qcut(cdystonia.age, 4)[:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, one can specify custom quantiles to act as cut points:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "quantiles = pd.qcut(segments.seg_length, [0, 0.01, 0.05, 0.95, 0.99, 1])\n",
    "quantiles[:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that you can easily combine discretiztion with the generation of indicator variables shown above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.get_dummies(quantiles).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Permutation and sampling\n",
    "\n",
    "For some data analysis tasks, such as simulation, we need to be able to randomly reorder our data, or draw random values from it. Calling NumPy's `permutation` function with the length of the sequence you want to permute generates an array with a permuted sequence of integers, which can be used to re-order the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "new_order = np.random.permutation(len(segments))\n",
    "new_order[:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using this sequence as an argument to the `take` method results in a reordered DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "segments.take(new_order).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare this ordering with the original:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "segments.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For random sampling, `DataFrame` and `Series` objects have a `sample` method that can be used to draw samples, with or without replacement:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vessels.sample(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vessels.sample(n=10, replace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data aggregation and GroupBy operations\n",
    "\n",
    "One of the most powerful features of Pandas is its **GroupBy** functionality. On occasion we may want to perform operations on *groups* of observations within a dataset. For exmaple:\n",
    "\n",
    "* **aggregation**, such as computing the sum of mean of each group, which involves applying a function to each group and returning the aggregated results\n",
    "* **slicing** the DataFrame into groups and then doing something with the resulting slices (*e.g.* plotting)\n",
    "* group-wise **transformation**, such as standardization/normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_grouped = cdystonia.groupby(cdystonia.patient)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This *grouped* dataset is hard to visualize\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, the grouping is only an intermediate step; for example, we may want to **iterate** over each of the patient groups:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for patient, group in cdystonia_grouped:\n",
    "    print('patient', patient)\n",
    "    print('group', group)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A common data analysis procedure is the **split-apply-combine** operation, which groups subsets of data together, applies a function to each of the groups, then recombines them into a new data table.\n",
    "\n",
    "For example, we may want to aggregate our data with with some function.\n",
    "\n",
    "![split-apply-combine](http://f.cl.ly/items/0s0Z252j0X0c3k3P1M47/Screen%20Shot%202013-06-02%20at%203.04.04%20PM.png)\n",
    "\n",
    "<div align=\"right\">*(figure taken from \"Python for Data Analysis\", p.251)*</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can aggregate in Pandas using the `aggregate` (or `agg`, for short) method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_grouped.agg(np.mean).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that the `treat` and `sex` variables are not included in the aggregation. Since it does not make sense to aggregate non-string variables, these columns are simply ignored by the method.\n",
    "\n",
    "Some aggregation functions are so common that Pandas has a convenience method for them, such as `mean`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_grouped.mean().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `add_prefix` and `add_suffix` methods can be used to give the columns of the resulting table labels that reflect the transformation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_grouped.mean().add_suffix('_mean').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# The median of the `twstrs` variable\n",
    "cdystonia_grouped['twstrs'].quantile(0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we wish, we can easily aggregate according to multiple keys:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia.groupby(['week','site']).mean().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternately, we can **transform** the data, using a function of our choice with the `transform` method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "normalize = lambda x: (x - x.mean())/x.std()\n",
    "\n",
    "cdystonia_grouped.transform(normalize).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is easy to do column selection within `groupby` operations, if we are only interested split-apply-combine operations on a subset of columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia_grouped['twstrs'].mean().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# This gives the same result as a DataFrame\n",
    "cdystonia_grouped[['twstrs']].mean().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you simply want to divide your DataFrame into chunks for later use, its easy to convert them into a dict so that they can be easily indexed out as needed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "chunks = dict(list(cdystonia_grouped))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "chunks[4]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default, `groupby` groups by row, but we can specify the `axis` argument to change this. For example, we can group our columns by `dtype` this way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "grouped_by_type = cdystonia.groupby(cdystonia.dtypes, axis=1)\n",
    "{g:grouped_by_type.get_group(g) for g in grouped_by_type.groups}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Its also possible to group by one or more levels of a hierarchical index. Recall `cdystonia2`, which we created with a hierarchical index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cdystonia2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cdystonia2.groupby(level='obs', axis=0)['twstrs'].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply\n",
    "\n",
    "We can generalize the split-apply-combine methodology by using `apply` function. This allows us to invoke any function we wish on a grouped dataset and recombine them into a DataFrame."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below takes a DataFrame and a column name, sorts by the column, and takes the `n` largest values of that column. We can use this with `apply` to return the largest values from every group in a DataFrame in a single call. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def top(df, column, n=5):\n",
    "    return df.sort_values(by=column, ascending=False)[:n]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see this in action, consider the vessel transit segments dataset (which we merged with the vessel information to yield `segments_merged`). Say we wanted to return the 3 longest segments travelled by each ship:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "top3segments = segments_merged.groupby('mmsi').apply(top, column='seg_length', n=3)[['names', 'seg_length']]\n",
    "top3segments.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that additional arguments for the applied function can be passed via `apply` after the function name. It assumes that the DataFrame is the first argument."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recall the microbiome data sets that we used previously for the concatenation example. Suppose that we wish to aggregate the data at a higher biological classification than genus. For example, we can identify samples down to *class*, which is the 3rd level of organization in each index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mb1.index[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the string methods `split` and `join` we can create an index that just uses the first three classifications: domain, phylum and class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class_index = mb1.index.map(lambda x: ' '.join(x.split(' ')[:3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mb_class = mb1.copy()\n",
    "mb_class.index = class_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, since there are multiple taxonomic units with the same class, our index is no longer unique:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mb_class.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can re-establish a unique index by summing all rows with the same class, using `groupby`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mb_class.groupby(level=0).sum().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise\n",
    "\n",
    "Load the dataset in `titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename='Data/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Women and children first?\n",
    "\n",
    "1. Describe each attribute, both with basic statistics and plots. State clearly your assumptions and discuss your findings.\n",
    "2. Use the `groupby` method to calculate the proportion of passengers that survived by sex.\n",
    "3. Calculate the same proportion, but by class and sex.\n",
    "4. Create age categories: children (under 14 years), adolescents (14-20), adult (21-64), and senior(65+), and calculate survival proportions by age category, class and sex."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "[Python for Data Analysis](http://shop.oreilly.com/product/0636920023784.do) Wes McKinney"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  },
  "nav_menu": {},
  "toc": {
   "navigate_menu": true,
   "number_sections": false,
   "sideBar": false,
   "threshold": "3",
   "toc_cell": true,
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
